<html>
<head>
    <meta content="width=device-width, initial-scale=1" charset="utf-8" name="viewport" />
    <title>SpliceAI Lookup</title>
    <link rel="shortcut icon" href="/icon.png">

    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.css" />
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.js"></script>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/underscore@1.12.0/underscore-min.js"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-172722632-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());
      gtag('config', 'UA-172722632-1');
    </script>
    <style>
        body {
            overflow-x: scroll;
        }
        body, .ui.form, label {
            font-size: 11pt !important;
        }
        td {
            padding: 5px 15px;
        }
        .upperBorder {
            border-top: 1px solid black;
        }
        table {
            font-size: 11pt !important;
            border-collapse: collapse;
            border-left: 0px !important;
            border-right: 0px !important;
        }
        .smallLink {
            font-size: 10pt;
            margin-left: 5px;
            margin-top: 3px;
        }

        .only-large-screen {
            display: none !important;
        }

        .transcript-color-legend {
            border-style:solid;
            min-width:23px;
            display:inline-block;
            margin-left: 20px;
        }

        @media screen and (min-width: 770px) {
            .only-large-screen {
                display: block !important;
            }
        }

        .results-div {
            width: 100%;
            position: static;
        }

        @media screen and (max-width: 1400px) {
            .results-div {
                width: 85vw;
                position: relative;
            }
        }

    </style>
</head>
<body>
    <div class="ui stackable grid">
      <div class="row">
        <div class="one wide column only-large-screen"></div>
        <div class="ten wide column">
            <div class="ui grid">
                <div class="row">
                    <div class="twelve wide column">
                        <div style="padding:25px 0px">
                            <div>
                                <div style="display:inline-block; min-width: 370px; ">Enter a variant below to see its <a href="https://www.cell.com/cell/pdf/S0092-8674(18)31629-5.pdf" target="_blank">SpliceAI</a> scores. </div>
                                    <a id="more-details1-button" href="#" onclick="$('#more-details1').show();$('#more-details1-button').hide();">[more details]</a>
                                <div id="more-details1" style="display:none">
                                    <br />
                                    This web interface and the underlying <a href="https://spliceailookup-api.broadinstitute.org/" target="_blank">SpliceAI server</a> are developed and maintained by the <a href="https://the-tgg.org/" target="_blank">TGG</a> at the <a href="https://www.broadinstitute.org/" target="_blank">Broad Institute</a>.<br />
                                    The <a href="https://spliceailookup-api.broadinstitute.org/" target="_blank">SpliceAI server</a> optionally uses pre-computed
                                    scores provided by Illumina for SNVs and small InDels, but runs the <a href="https://github.com/Illumina/SpliceAI" target="_blank">SpliceAI model</a>
                                    directly for scores not available in the pre-computed files - such as for larger InDels, non-default "Max distance" values, etc.<br />
				    <br />
				    NOTE: The server supports not more than several queries per-minute per-user. If you need to batch-process many variants, please use the
				    <a href="https://github.com/broadinstitute/SpliceAI-lookup/blob/master/README.md#local-install" target="_blank">source code</a>
				    to set up your own local instance of the API server or just run the underlying SpliceAI model directly.<br />
                                    <br />
                                    For more information about the SpliceAI model, see <a href="https://doi.org/10.1016/j.cell.2018.12.015" target="_blank">Jaganathan et al, Cell 2019</a> and <a href="https://github.com/Illumina/SpliceAI" target="_blank">github.com/Illumina/SpliceAI</a>,
                                    and this <a href="https://www.youtube.com/watch?v=oJvhj-tYbBI" target="_blank">recorded talk</a> by Kishore Jaganathan.<br />
                                    <br />
				    Since February 2023, <a href="https://github.com/tkzeng/Pangolin" target="_blank">Pangolin scores</a> from [<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02664-4" target="_blank">Zeng & Li 2022</a>] are also shown below the SpliceAI results.
                                </div>
                            </div>
                            <br />
                            <div style="padding-left: 0px; color: gray">
                                <table>
                                    <thead style="text-align: left">
                                        <tr>
                                            <th style="padding-bottom: 20px; min-width:370px">Examples &nbsp; (on hg38): </th>
                                            <th></th>
                                        </tr>
                                    </thead>
                                    <tbody>
                                        <tr>
                                            <td>chr8-140300616-T-G</td>
                                            <td></td>
                                        </tr>
                                        <tr>
                                            <td>chr8 &nbsp; 140300616 &nbsp; T &nbsp; G</td>
                                            <td></td>
                                        </tr>
                                        <tr>
                                            <td style="padding-right: 0px">NM_000552.4(VWF):c.3797C>A (p.Pro1266Gln)</td>
                                            <td style="padding-left: 0px">
                                                <a id="more-details2-button" href="#" onclick="$('.more-details2').show();$('#more-details2-button').hide();">[show more examples]</a>
                                            </td>
                                        </tr>
                                        <tr style="display:none;" class="more-details2">
                                            <td>8:140300616 T&gt;G</td>
                                            <td>'chr' <i>prefix is optional</i></td>
                                        </tr>
                                        <tr style="display:none;" class="more-details2">
                                            <td>1:930130:C:G </td>
                                            <td><i>position with overlapping genes</i></td>
                                        </tr>
                                        <tr style="display:none;" class="more-details2">
                                            <td>1-1042601-A-AGAGAG </td>
                                            <td><i>insertion of GAGAG</i></td>
                                        </tr>
                                        <tr style="display:none;" class="more-details2">
                                            <td>1-1042466-GGGC-G</td>
                                            <td><i>deletion of GGC</i></td>
                                        </tr>
                                        <tr style="display:none;" class="more-details2">
                                            <td>NM_000249.4(MLH1):c.116G>A</td>
                                            <td><i>another HGVS exapmle</i></td>
                                        </tr>
                                    </tbody>
                                </table>
                            </div>
                        </div>
<!--
                        <br />
                        <div style="font-size: large; color: darkred">
                          <br />
                          <b>The server is currently down for updates</b><br />
                          <br />
                          Please check back in ~1/2 hour.
                        </div>
                        <br />
                        <br />
-->			
                        <div class="ui form" style="width:100%">
                            <div class="ui input" style="padding-bottom:30px; padding-right:8px; width:100%">
                                <input id="search-box" type="text" style="width: 100%; padding: 7px 10px" placeholder="Enter a variant...">
                            </div>
                            <!-- div class="field" style="padding-bottom: 10px;">
                                <textarea id="search-box2" rows="1" style="width: 100%" placeholder="Enter variant..."></textarea>
                            </div -->
                            <div class="ui stackable grid">
                                <div class="row">
                                    <div class="thirteen wide column">
                                        <div class="field" style="padding-right:10px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Genome version:</b></span>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="hg" value="37" /><label>hg19</label>
                                            </div>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="hg" value="38" checked /><label>hg38</label>
                                            </div>
                                        </div>
                                        <div class="field" style="padding-right:10px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Score type:</b></span>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="mask" value="1" /><label>masked</label>
                                            </div>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="mask" value="0" checked /><label>raw</label>
                                            </div>
                                            <i style="margin-left:20px" class="question circle outline icon" data-position="right center" data-content="Splicing changes corresponding to strengthening annotated splice sites and weakening unannotated splice sites are typically much less pathogenic than weakening annotated splice sites and strengthening unannotated splice sites. Selecting 'masked' will hide the score for such splicing changes and show 0 instead. Selecting 'raw' will show all scores. SpliceAI developers recommend using 'raw' scores for alternative splicing analysis and 'masked' scores for variant interpretation."></i>
                                        </div>
                                        <div class="field" style="padding-right:10px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Max distance:</b></span>
                                            <div class="ui input" style="padding-left:15px; padding-bottom:10px; padding-right:8px; width:100px">
                                                <input id="max-distance-input" type="text" value="50" style="padding:7px 10px">
                                            </div>
                                            <i class="question circle outline icon" data-position="right center" data-content="For each variant, SpliceAI looks within a window (+/- 50bp by default) to see how the variant affects the probabilities of different positions being splice acceptors or donors. The distance specified here controls the size of this window. The maximum allowed value is 10,000bp."></i>
                                        </div>
                                        <div class="field" style="padding-right:10px; padding-bottom:12px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Use Illumina's pre-computed scores:</b></span>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="use-precomputed-scores" value="1" /><label>yes</label>
                                            </div>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="use-precomputed-scores" value="0" checked /><label>no</label>
                                            </div>
                                            <i style="margin-left:20px" class="question circle outline icon" data-position="right center" data-content="Illumina provides pre-computed SpliceAI scores for all SNPs and small InDels within Gencode v24 canonical transcripts. If 'yes' is selected here, the server will quickly check these pre-computed files to see if it can avoid the relatively slow step of running the SpliceAI model. Even with 'yes' selected, it will still run the model if it finds that the pre-computed files don't contain the variant. Selecting 'no' will cause the server to always skip the pre-computed files and run the model. The benefit of skipping the pre-computed files is that, although the scores may take longer, they will always be computed for Gencode v38 transcripts and will include ENSG and ENST ids in addition to the gene name."></i>
                                        </div>
                                    </div>
                                    <div class="three wide column">
                                        <button id="submit-button" class="ui primary button" style="margin-bottom: 15px">Submit</button>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        </div>
        <div class="four wide column">
            <div class="only-large-screen" style="padding:25px 0px">
                <div>
                    <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues" target="_blank">SpliceAI-lookup/issues</a>: issues or feature requests for this website.
                    <a href="https://github.com/illumina/SpliceAI" target="_blank">SpliceAI/issues</a>: issues with the underlying SpliceAI model.<br />
                    <br />
                    <div><i>What's New</i></div>
                    <br />
                    <b>May 12, 2023</b><br />
                    - added REF & ALT score columns to show SpliceAI's underlying predictions for each haplotype (using <a href="https://github.com/Illumina/SpliceAI/pull/92">SpliceAI PR by @h-joshi</a>).
                    The &#916;&nbsp; score column is the difference of these two scores.<br />
                    <br />		    
                    <b>May 10, 2023</b><br />
                    - added support for MNPs (see <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues/1">issue #1</a> & <a href="https://github.com/Illumina/SpliceAI/pull/119">SpliceAI PR by @kdahlo</a>)<br />
                    - fixed bug where "raw" Pangolin scores were shown regardless of whether the user selected "raw" or "masked".<br />
                    - updated to <a href="https://www.gencodegenes.org/human/releases.html">Gencode v43</a> (was previously on v42)<br />
		            <br />

                    <a id="more-details3-button" href="#" onclick="$('#more-details3').show();$('#more-details3-button').hide();">[show older updates]</a>
                    <div id="more-details3" style="display:none">

                        <b>Feb 5, 2023</b><br />
                        - show <a href="https://github.com/tkzeng/Pangolin">Pangolin scores</a> in addition to SpliceAI scores.<br />
                        <br />
                        <b>June 2, 2021</b><br />
                        - show RefSeq ids for the subset of Ensembl ENST ids (~30%) that have one or more matching RefSeq NM ids according to Ensembl. See <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues/8">issue #8</a> for details.<br />
                        <br />
                        <b>April 11, 2021</b><br />
                        - show gray background for non-coding transcripts<br />
                        - switch to showing all non-coding transcripts. (Previously, transcripts with Gencode biotypes
                          like <i>lncRNA</i>, <i>processed_pseudogene</i>, <i>processed_transcript</i>, <i>retained_intron</i>, and
                          <i>nonsense_mediated_decay</i> were filtered out if they overlapped <i>protein_coding</i> transcripts).
                          See <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues/6#issuecomment-816942824">issue #6</a> for details.<br />
                    </div>
                </div>
                <br />
                <div>
                    <b>Related web tools:</b><br />
                    <a href="https://liftover.broadinstitute.org">liftover</a>: for variants/positions/intervals (hg19 <=> hg38 <=> T2T)<br />
                    <a href="https://cma-search.broadinstitute.org">CMA search</a>: search OMIM by interval, gene or phenotype<br />
                </div>
            </div>
        </div>
        <div class="one wide column only-large-screen"></div>
      </div>
      <div class="row">
          <div class="one wide column only-large-screen"></div>
          <div class="fourteen wide column">
              <div class="results-div">
                  <div id="error-box" style="color:darkred"></div>
                  <div id="response-box"></div>
              </div>
          </div>
          <div class="one wide column only-large-screen"></div>
      </div>
    </div>

    <script>
      const VARIANT_RE = new RegExp(
	"^[\\s]*" +
        "(chr)?([0-9XYMTt]{1,2})" +
        "[-\\p{Pd}\\s:]+" +
        "([0-9,]+)" +
        "[-\\p{Pd}\\s:]*" +
        "([ACGT]+)" +
        "[-\\p{Pd}\\s:>]+" +
        "([ACGT]+)",
        'iu'
      )

      const GRCH37_contigs = {
        "1": "NC_000001.10",
        "2": "NC_000002.11",
        "3": "NC_000003.11",
        "4": "NC_000004.11",
        "5": "NC_000005.9",
        "6": "NC_000006.11",
        "7": "NC_000007.13",
        "8": "NC_000008.10",
        "9": "NC_000009.11",
        "10": "NC_000010.10",
        "11": "NC_000011.9",
        "12": "NC_000012.11",
        "13": "NC_000013.10",
        "14": "NC_000014.8",
        "15": "NC_000015.9",
        "16": "NC_000016.9",
        "17": "NC_000017.10",
        "18": "NC_000018.9",
        "19": "NC_000019.9",
        "20": "NC_000020.10",
        "21": "NC_000021.8",
        "22": "NC_000022.10",
        "X": "NC_000023.10",
        "Y": "NC_000024.9",
        "M":  "NC_012920.1",
        "MT": "NC_012920.1",
      }

      const GRCH38_contigs = {
        "1" : "NC_000001.11",
        "2" : "NC_000002.12",
        "3" : "NC_000003.12",
        "4" : "NC_000004.12",
        "5" : "NC_000005.10",
        "6" : "NC_000006.12",
        "7" : "NC_000007.14",
        "8" : "NC_000008.11",
        "9" : "NC_000009.12",
        "10": "NC_000010.11",
        "11": "NC_000011.10",
        "12": "NC_000012.12",
        "13": "NC_000013.11",
        "14": "NC_000014.9",
        "15": "NC_000015.10",
        "16": "NC_000016.10",
        "17": "NC_000017.11",
        "18": "NC_000018.10",
        "19": "NC_000019.10",
        "20": "NC_000020.11",
        "21": "NC_000021.9",
        "22": "NC_000022.11",
        "X" : "NC_000023.11",
        "Y" : "NC_000024.10",
        "M":  "NC_012920.1",
        "MT": "NC_012920.1",
      }

      const SPLICEAI_SCORE_FIELDS = "SYMBOL|DS_AG|DS_AL|DS_DG|DS_DL|DP_AG|DP_AL|DP_DG|DP_DL|DS_AG_REF|DS_AG_ALT|DS_AL_REF|DS_AL_ALT|DS_DG_REF|DS_DG_ALT|DS_DL_REF|DS_DL_ALT".split("|")
      const PANGOLIN_SCORE_FIELDS = "SYMBOL|DS_SG|DS_SL|DP_SG|DP_SL".split("|")  // splice_gain_score, splice_loss_score, splice_gain_pos, splice_loss_pos

      const CANONICAL_TRANSCRIPT_BACKGROUND_COLOR = "#eaf3ff"
      const CODING_TRANSCRIPT_BACKGROUND_COLOR = "white"
      const NON_CODING_TRANSCRIPT_BACKGROUND_COLOR = "#e5e5e5"

      const RED_BACKGROUND = "#fccfb8"
      const YELLOW_BACKGROUND = "#fff19d"
      const GREEN_BACKGROUND = "#cdffd7"

      function parseVariant(variant) {
        const m = variant.match(VARIANT_RE)
        if (m) {
          const chrom = m[2].toUpperCase()
          return {
            chrom: chrom,
            pos: m[3].replace(/,/g, ""),
            ref: m[4],
            alt: m[5],
          }
        }

        return null
      }


      function parseVariantToHGVS(variant, genome_version) {
        const m = variant.match(VARIANT_RE)
        if (m) {
          let chrom = m[2].toUpperCase()
          if (genome_version == "37") {
            chrom = GRCH37_contigs[chrom]
          } else if(genome_version == "38") {
            chrom = GRCH38_contigs[chrom]
          } else {
            throw "Invalid genome_version: " + genome_version
          }

          const pos = parseInt(m[3].replace(/,/g, ""))
          const ref = m[4]
          const alt = m[5]

          if (ref.length > 1 && alt.length > 1) {
            variant = chrom + ":g." + pos + "_" + (pos + ref.length - 1) + "delins" + alt
          } else if (ref.length == alt.length) {
            variant = chrom + ":g." + pos + ref + ">" + alt
          } else if (ref.length > alt.length) {
            //deletion (https://varnomen.hgvs.org/recommendations/DNA/variant/deletion/)
            variant = chrom + ":g." + (pos + 1) + "_" + (pos + ref.length - alt.length) + "del"
          } else if (ref.length < alt.length) {
            //insertion (https://varnomen.hgvs.org/recommendations/DNA/variant/insertion/)
            variant = chrom + ":g." + pos + "_" + (pos + 1) + "ins" + alt.slice(1)
          }

          console.log("Parsed variant", variant)
        }

        return variant
      }


      function getScoreStyle(score) {

        score = parseFloat(score)
        if(score >= 0.8) {
          return "style='background-color:"+RED_BACKGROUND+"'"
        } else if (score >= 0.5) {
          return "style='background-color:"+YELLOW_BACKGROUND+"'"
        } else if (score >= 0.2) {
          return "style='background-color:"+GREEN_BACKGROUND+"'"
        } else if (score == 0) {
          return "style='color:#BBBBBB'"
        }
      }

      function getUCSCBrowserUrl(genome_version, chrom, pos) {

        genome_version = genome_version.replace('37', '19')
        chrom = chrom.toUpperCase().replace('CHR', '')
        return "https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg" + genome_version + "&position=chr" + chrom + ":" + pos
      }

      function generateResultsTable(responses) {

        let parsed_scores_or_errors = []
        responses.filter(r => r).forEach(function(response) {
            const is_pangolin = response.source && response.source.startsWith("pangolin")
            if (response.error) {
                parsed_scores_or_errors.push({
                    'error': response.error,
                    'source': response.source,
                    'is_pangolin': is_pangolin,
                    'hg': response.hg,
                    'variant': response.variant,
                    'sort_key': 0,
                })
                return
            }

            response.scores.forEach(function(scores) {
                const splice_prediction_values = scores.split("|")
                const splice_prediction_dict = {};

                (is_pangolin ? PANGOLIN_SCORE_FIELDS : SPLICEAI_SCORE_FIELDS).forEach(function(k, i) {
                    splice_prediction_dict[k] = splice_prediction_values[i]
                    if (k.startsWith("DS_")) {
                        splice_prediction_dict[k] = Math.abs(parseFloat(splice_prediction_dict[k]))
                    }
                })

                const score_types = is_pangolin ? ["SG", "SL"] : ["DG", "DL", "AG", "AL"]
                score_types.forEach(function(score_type) {
                    try {
                        splice_prediction_dict[score_type + "_url"] = getUCSCBrowserUrl(response.hg, response.chrom, response.pos + parseInt(splice_prediction_dict["DP_" + score_type]))
                    } catch(e) {
                        splice_prediction_dict[score_type + "_url"] = "#"
                    }
                })

                try {
                    splice_prediction_dict["url"] = getUCSCBrowserUrl(response.hg, response.chrom, response.pos)
                } catch(e) {
                    splice_prediction_dict["url"] = "#"
                }

                splice_prediction_dict['source'] = response.source
                splice_prediction_dict['is_pangolin'] = is_pangolin
                splice_prediction_dict['hg'] = response.hg
                splice_prediction_dict['variant'] = response.variant

                // compute sort key
                const splice_prediction_tool = is_pangolin ? 1 : 0;

                const gene_name_fields = (splice_prediction_dict["SYMBOL"] || "").split("---")
                const gene_name = gene_name_fields[0]
                const transcript_id = gene_name_fields.length >= 3 && gene_name_fields[2]
                const is_canonical_transcript = (gene_name_fields.length >= 4 && gene_name_fields[3] === "yes") ? 0 : 1
                const transcript_type = gene_name_fields.length >= 5 && gene_name_fields[4]
                const is_protein_coding = !transcript_type || transcript_type == "protein_coding" ? 0 : 1
                const refseq_transcript_id_is_in_query = gene_name_fields.length >= 6 && gene_name_fields[5] && gene_name_fields[5].split(",").some(refseq_transcript_id => response.raw.includes(refseq_transcript_id))
                const contains_hgvs_transcript_id = response.raw.includes(transcript_id) || refseq_transcript_id_is_in_query ? 0 : 1;

                splice_prediction_dict['sort_key'] = `${splice_prediction_tool}${contains_hgvs_transcript_id}`
                splice_prediction_dict['sort_key'] += `${is_protein_coding}${is_canonical_transcript}`
                splice_prediction_dict['sort_key'] += `---${gene_name}`

                parsed_scores_or_errors.push(splice_prediction_dict)
            })
        })

        //sort by (gene name, is canonical transcript, is non-coding transcriipt)
        parsed_scores_or_errors = _.chain(parsed_scores_or_errors).sortBy('sort_key').value()

        console.log("Parsed scores:", parsed_scores_or_errors)

        return "<table class='ui stackable table'>" +
          "<tr>" +
              "<td><b>variant</b></td>" +
              "<td><b>gene</b><div style='float:right;'>" +
                `<span class='transcript-color-legend' style='background-color:${CANONICAL_TRANSCRIPT_BACKGROUND_COLOR}!important'>&nbsp;</span> = canonical transcript ` +
                `<span class='transcript-color-legend' style='background-color:${NON_CODING_TRANSCRIPT_BACKGROUND_COLOR}!important'>&nbsp;</span> = non-coding transcript` +
              "</td>" +
              "<td><b>&#916;&nbsp; type</b></td>" +
              "<td><b>&#916;&nbsp; score</b>  " + "<i class='question circle outline icon' data-content='Delta scores range from 0 to 1 and can be interpreted as the probability that the variant affects splicing at any position within a window around it (+/- 50bp by default). In the SpliceAI paper, a detailed characterization is provided for 0.2 (high recall), 0.5 (recommended), and 0.8 (high precision) cutoffs. Consequently, scores of 0.2 or higher appear in green, 0.5 or higher appear in yellow, and 0.8 or higher appear in red.'></i>"+"</td>" +
              "<td><b>pre-mRNA position</b>  "+ "<i class='question circle outline icon' data-content='For each variant, SpliceAI looks within a window (+/- 50bp by default) to see how the variant affects the probabilities of different positions in the pre-mRNA being splice acceptors or donors. The offsets in this column represent positions with the biggest change in probability within the window. Negative values are upstream (5&#039;) of the variant, and positive are downstream (3&#039;) of the variant.'></i>"+"</td>" +
              "<td><b>REF score</b>  "+ "<i class='question circle outline icon' data-content='Raw probability for the REF allele'></i>"+"</td>" +
              "<td><b>ALT score</b>  "+ "<i class='question circle outline icon' data-content='Raw probability for the ALT allele'></i>"+"</td>" +
          "</tr>" + parsed_scores_or_errors.map(function(scores_or_error) {
                if (scores_or_error.error) {
                    return "<tr>" +
                        "<td colspan='7'>" +
                            "<div style='color:darkred'>" + ( scores_or_error.is_pangolin ? 'Pangolin ' : 'SpliceAI ' ) + scores_or_error.error + "</div>" +
                        "</td>" +
                    "</tr>"
                }
                let scores = scores_or_error
                const variant_ucsc_url = scores.url
                const gnomAD_version = scores.hg === "38"? "gnomad_r3" : "gnomad_r2_1"
                const variant_gnomAD_url = "https://gnomad.broadinstitute.org/variant/" + scores.variant + "?dataset=" + gnomAD_version
                const gene_name_fields = (scores.SYMBOL || "").split("---")
                const gene_name = gene_name_fields[0]

                const is_canonical_transcript = gene_name_fields.length >= 4 && gene_name_fields[3] === "yes"
                const transcript_type = gene_name_fields.length >= 5 && gene_name_fields[4]
                const is_protein_coding = !transcript_type || transcript_type == "protein_coding"
                const refseq_transcript_ids = gene_name_fields.length >= 6 && gene_name_fields[5] && gene_name_fields[5].split(",")
                const background_color = !is_protein_coding ? NON_CODING_TRANSCRIPT_BACKGROUND_COLOR : (
                    is_canonical_transcript ? CANONICAL_TRANSCRIPT_BACKGROUND_COLOR : CODING_TRANSCRIPT_BACKGROUND_COLOR
                )

                let gene_and_transcript_id = ""
                if (gene_name_fields.length >= 3 ) {
                    const gene_id_link = `<a href="https://useast.ensembl.org/Homo_sapiens/Gene/Summary?g=${gene_name_fields[1].split(".")[0]}" target="_blank">${gene_name_fields[1]}</a>`
                    const transcript_id_link = ` / <a href="https://useast.ensembl.org/Homo_sapiens/Transcript/Summary?t=${gene_name_fields[2].split(".")[0]}" target="_blank">${gene_name_fields[2]}</a>`
                    const canonical_transcript_text = is_canonical_transcript ? '<div class="smallLink">canonical transcript</div>' : ''
                    const biotype_link = `<div class="smallLink"><a href="https://www.gencodegenes.org/pages/biotypes.html" target="_blank">biotype: ${transcript_type.replace(/_/g, " ")}</a></div>`
                    const refseq_transcript_id_link = refseq_transcript_ids ? (" /" + refseq_transcript_ids.map(refseq_transcript_id => ` <a href="https://www.ncbi.nlm.nih.gov/search/all/?term=${refseq_transcript_id}" target="_blank">${refseq_transcript_id}</a>`).join(', ')) : ''
                    gene_and_transcript_id = ` (${gene_id_link}${transcript_id_link}${refseq_transcript_id_link})<br /> ${biotype_link} ${canonical_transcript_text}`
                }
                //round to 2 decimal places
                if (scores.is_pangolin) {
                    scores.DS_SL = Math.round(scores.DS_SL * 100) / 100
                    scores.DS_SG = Math.round(scores.DS_SG * 100) / 100
                } else {
                    scores.DS_AG = Math.round(scores.DS_AG * 100) / 100
                    scores.DS_AL = Math.round(scores.DS_AL * 100) / 100
                    scores.DS_DL = Math.round(scores.DS_DL * 100) / 100
                    scores.DS_DG = Math.round(scores.DS_DG * 100) / 100
                    scores.DS_AL_REF = Math.round(scores.DS_AL_REF * 100) / 100
                    scores.DS_AL_ALT = Math.round(scores.DS_AL_ALT * 100) / 100
                    scores.DS_DL_REF = Math.round(scores.DS_DL_REF * 100) / 100
                    scores.DS_DL_ALT = Math.round(scores.DS_DL_ALT * 100) / 100
                    scores.DS_AG_REF = Math.round(scores.DS_AG_REF * 100) / 100
                    scores.DS_AG_ALT = Math.round(scores.DS_AG_ALT * 100) / 100
                    scores.DS_DG_REF = Math.round(scores.DS_DG_REF * 100) / 100
                    scores.DS_DG_ALT = Math.round(scores.DS_DG_ALT * 100) / 100
                }

                let result= "<tr class='upperBorder' style='background-color:" + background_color+"'>" +
                  "<td rowspan='" + (scores.is_pangolin ? 2: 4) + "' style='vertical-align: top'>" +
                  "<div style='margin-right:10px; display: inline-block'>" + scores.variant + "</div><br class='only-large-screen'/><br class='only-large-screen'/>" +
                    "<a href='"+variant_ucsc_url+"' class='smallLink' target='_blank'>UCSC</a>, " +
                    "<a href='"+variant_gnomAD_url+"' class='smallLink' target='_blank'>gnomAD</a>" +
                  "</td>"+
                  "<td rowspan='" + (scores.is_pangolin ? 2: 4) + "' style='vertical-align: top'>" +
                    "<div style='margin-right:10px; display: inline-block'>"+gene_name+(gene_and_transcript_id || '') +"</div><br class='only-large-screen'/><br class='only-large-screen'/>" +
                    "<a href='https://www.omim.org/search?search="+gene_name+"' class='smallLink' target='_blank'>OMIM</a>, " +
                    "<a href='https://gtexportal.org/home/gene/"+gene_name+"' class='smallLink' target='_blank'>GTEx</a>, " +
                    "<a href='https://gnomad.broadinstitute.org/gene/"+gene_name+"?dataset="+gnomAD_version+"' class='smallLink' target='_blank'>gnomAD</a>, " +
                    "<a href='https://search.clinicalgenome.org/kb/genes?page=1&size=25&search="+gene_name+"' class='smallLink' target='_blank'>ClinGen</a>, <br class='only-large-screen'/>" +
                    "<a href='https://useast.ensembl.org/Homo_sapiens/Gene/Summary?g="+gene_name+"' class='smallLink' target='_blank'>Ensembl</a>, " +
                    "<a href='https://decipher.sanger.ac.uk/gene/"+gene_name+"' class='smallLink' target='_blank'>Decipher</a>, " +
                    "<a href='https://www.genecards.org/cgi-bin/carddisp.pl?gene="+gene_name+"' class='smallLink' target='_blank'>GeneCards</a>" +
                  "</td>" +
                    (scores.is_pangolin ?
                            "<td>Splice Loss</td><td "+getScoreStyle(scores.DS_SL)+">" + scores.DS_SL.toFixed(2) +"</td><td>" + (scores.DS_SL == 0 ? "" : scores.DP_SL+" bp") + "</td>"+
                            "</tr><tr style='background-color:" + background_color+"'>"+
                            "<td>Splice Gain</td><td "+getScoreStyle(scores.DS_SG)+">" + scores.DS_SG.toFixed(2) +"</td><td>" + (scores.DS_SG == 0 ? "" : scores.DP_SG+" bp") + "</td>"
                        :
                            "<td>Acceptor Loss</td><td "+getScoreStyle(scores.DS_AL)+">" + scores.DS_AL.toFixed(2) +"</td><td>" + ((scores.DS_AL == 0 && scores.DS_AL_REF == 0 && scores.DS_AL_ALT == 0)? "" : scores.DP_AL+" bp") + "</td><td "+getScoreStyle(scores.DS_AL_REF)+">" + scores.DS_AL_REF.toFixed(2) + "</td><td "+getScoreStyle(scores.DS_AL_ALT)+">" + scores.DS_AL_ALT.toFixed(2) + "</td>"+
                            "</tr><tr style='background-color:" + background_color+"'>"+
                            "<td>Donor Loss</td><td "+getScoreStyle(scores.DS_DL)+">" + scores.DS_DL.toFixed(2) +"</td><td>" + ((scores.DS_DL == 0 && scores.DS_DL_REF == 0 && scores.DS_DL_ALT == 0)? "" : scores.DP_DL+" bp") + "</td><td "+getScoreStyle(scores.DS_DL_REF)+">" + scores.DS_DL_REF.toFixed(2) + "</td><td "+getScoreStyle(scores.DS_DL_ALT)+">" + scores.DS_DL_ALT.toFixed(2) + "</td>"+
                            "</tr><tr style='background-color:" + background_color+"'>"+
                            "<td>Acceptor Gain</td><td "+getScoreStyle(scores.DS_AG)+">" + scores.DS_AG.toFixed(2) +"</td><td>" + ((scores.DS_AG == 0 && scores.DS_AG_REF == 0 && scores.DS_AG_ALT == 0)? "" : scores.DP_AG+" bp") + "</td><td "+getScoreStyle(scores.DS_AG_REF)+">" + scores.DS_AG_REF.toFixed(2) + "</td><td "+getScoreStyle(scores.DS_AG_ALT)+">" + scores.DS_AG_ALT.toFixed(2) + "</td>"+
                            "</tr><tr style='background-color:" + background_color+"'>"+
                            "<td>Donor Gain</td><td "+getScoreStyle(scores.DS_DG)+">" + scores.DS_DG.toFixed(2) +"</td><td>" + ((scores.DS_DG == 0 && scores.DS_DG_REF == 0 && scores.DS_DG_ALT == 0)? "" : scores.DP_DG+" bp") + "</td><td "+getScoreStyle(scores.DS_DG_REF)+">" + scores.DS_DG_REF.toFixed(2) + "</td><td "+getScoreStyle(scores.DS_DG_ALT)+">" + scores.DS_DG_ALT.toFixed(2) + "</td>"
                    ) +
                  "</tr>"

                if(scores.is_pangolin) {
                    result = "<tr><td style='white-space: nowrap'><b>Pangolin scores: <a href='https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02664-4' target='_blank'><i class='question circle outline icon' data-position='right center' data-content='Pangolin scores are computed on the server by running the Pangolin model with the user-selected parameters (ie. max distance).'></i></a></b> </td><td></td><td></td><td></td><td></td></tr>" + result
                }
                return result
              }).join("") +
          "</table>"
      }

      $.urlParam = function (name) {
        const results = new RegExp('[\?&#]?' + name + '=([^&#]*)').exec(window.location.hash);
        return (results !== null) ? decodeURIComponent(results[1]) || 0 : false;
      }

      $(document).ready(function() {
        $('.ui.radio.checkbox').checkbox()
        $('.question').popup()
        $("#search-box").focus()

        $("#search-box,#max-distance-input").keydown(function (event) {
          if ((event.keyCode || event.which) == 13) {
            $("#submit-button").click()
          }
        });

        $("#submit-button").click(function() {
          const variant = $("#search-box").val().trim()
          if (!variant) {
            return
          }

          const genome_version = $("input[name='hg']:checked").val().trim()
          const mask = $("input[name='mask']:checked").val().trim()
          const max_distance = $("#max-distance-input").val().trim()
          const use_precomputed_scores = $("input[name='use-precomputed-scores']:checked").val().trim()
          const variant_hgvs = parseVariantToHGVS(variant, genome_version)

          $("#error-box").hide()
          $("#response-box").hide()

          $("#submit-button").addClass(["loading", "disabled"])

          let ensembl_api_url_prefix = ""
          if (genome_version == "37") {
            ensembl_api_url_prefix = "grch37."
          }

          $.getJSON(
            "https://" + ensembl_api_url_prefix + "rest.ensembl.org/vep/human/hgvs/" + variant_hgvs + "?content-type=application/json&vcf_string=1"
          ).catch(function(error) {
            let errorText = error.responseText
            try {
              errorText = JSON.parse(errorText).error
            } catch(e) {
              console.warn(e)
            }

            const refAlleleError = errorText.match(new RegExp("[(]([ACGTRYSWKMBDHVN]+)[)] does not match reference allele given by HGVS notation"))
            if (refAlleleError) {
                errorText = variant + " has unexpected reference allele. The hg" + genome_version + " reference allele should be: " + refAlleleError[1]
            }

            errorText = "Ensembl API error: " + errorText
            console.warn(errorText)
            throw { responseText: errorText }
          }).then(function(response) {
            console.log("Esnembl API response", response)
            if (!response || !response[0]) {
              throw { responseText: "Unable to parse Ensembl API repsonse: " + response }
            }

            const vcf_string = (response[0].vcf_string || "---").split("-")
            const chrom = vcf_string[0]
            const pos = vcf_string[1]
            const ref = vcf_string[2]
            const alt = vcf_string[3]
            const most_severe_consequence = (response[0].most_severe_consequence || "").replace(/_/g, " ")

            console.log("chrom:", chrom, "pos:", pos, "ref:", ref, "alt:", alt)
            console.log("Most severe consequence:", most_severe_consequence)

            // get SpliceAI and Pangolin scores from API endpoints in parallel
            const base_api_url = location.hostname == "" ? "http://localhost:8080" : "https://spliceailookup-api.broadinstitute.org"
            //const base_api_url = "https://spliceailookup-api.broadinstitute.org"
            const getJSON_promises = ["spliceai", "pangolin"].map((endpoint) => {
                const promise = $.getJSON(`${base_api_url}/${endpoint}/`,
                    {
                        hg: genome_version,
                        distance: max_distance,
                        mask: mask,
                        precomputed: use_precomputed_scores,
                        variant: chrom + '-' + pos + '-' + ref + '-' + alt,
                        raw: variant,
                    }
                )
                return promise
            })

            return Promise.all(getJSON_promises.map(p => p.catch(e => console.log(e))))

          }).then(function(response) {
              console.log("spliceai API response was:", response[0])
              console.log("pangolin API response was:", response[1])

              $("#error-box").hide()
              $("#response-box").html(generateResultsTable(response))
              $("#response-box").show()

              // set url
              window.location.hash = "#" + $.param({
                  variant: variant,
                  hg: genome_version,
                  distance: max_distance,
                  mask: mask,
                  precomputed: use_precomputed_scores,
              })

          }).catch(function(error) {

            $("#response-box").hide()
            let errorText = error.responseText
            try {
              errorText = JSON.parse(errorText).error
            } catch(e) {
            }

            if (!errorText) {
                errorText = "ERROR: Unable to reach the SpliceAI-lookup API server. Either the API server is down or your internet isn't working."
            }
            console.warn(errorText)
            $("#error-box").html(errorText)
            $("#error-box").show()
          }).always(function() {
            $("#submit-button").removeClass(["loading", "disabled"]);
            $('.question').popup()
          })
        })

        const hgFromUrl = $.urlParam('hg')
        if (hgFromUrl) {
          $("input[name='hg'][value='"+hgFromUrl+"']").prop("checked", true)
        }

        const maxDistanceFromUrl = $.urlParam('distance')
        if (maxDistanceFromUrl) {
            $("#max-distance-input").val(maxDistanceFromUrl)
        }

        const maskFromUrl = $.urlParam('mask')
        if (maskFromUrl) {
          $("input[name='mask'][value='"+maskFromUrl+"']").prop("checked", true)
        }

        const precomputedFromUrl = $.urlParam('precomputed')
        if (precomputedFromUrl) {
          $("input[name='use-precomputed-scores'][value='"+precomputedFromUrl+"']").prop("checked", true)
        }

        const variantFromUrl = $.urlParam('variant')
        if (variantFromUrl) {
          $("#search-box").val(variantFromUrl)
          $("#submit-button").click()
        }
      })

    </script>
</body>
</html>
