<html>
<head>
    <script>
        //let baseApiUrl = "https://spliceailookup-api.broadinstitute.org"
        //if (location.origin && location.origin.startsWith("http") && !location.origin.includes("broadinstitute.org")) {
        //    baseApiUrl = location.origin
        //}

        const baseApiUrl = {
            "normalize": "https://liftover-xwkwwwxdwq-uc.a.run.app",
            "pangolin-37": "https://pangolin-37-xwkwwwxdwq-uc.a.run.app",
            "pangolin-38": "https://pangolin-38-xwkwwwxdwq-uc.a.run.app",
            "spliceai-37": "https://spliceai-37-xwkwwwxdwq-uc.a.run.app",
            "spliceai-38": "https://spliceai-38-xwkwwwxdwq-uc.a.run.app",
        }

        //console.log("Using SpliceAI API url:", baseApiUrl)
    </script>
    <meta content="width=device-width, initial-scale=1" charset="utf-8" name="viewport"/>
    <title>SpliceAI Lookup</title>
    <link rel="shortcut icon" href="/icon.png">

    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.css"/>
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.js"></script>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/underscore@1.12.0/underscore-min.js"></script>
    <script type="text/javascript" src="https://broadinstitute.github.io/SpliceAI-lookup-dev/tabix-bundle.js"></script>
    <script type="text/javascript" src="https://storage.googleapis.com/tgg-viewer/igv.js-bw2/dist/igv.min.js?version=2024-08-06"></script>
    <!-- script type="text/javascript" src="igv.min.js"></script -->

    <script async src="https://www.googletagmanager.com/gtag/js?id=G-483HVKZCL1"></script>
    <script>
        window.dataLayer = window.dataLayer || [];
        function gtag(){dataLayer.push(arguments);}
        gtag('js', new Date());
        gtag('config', 'G-483HVKZCL1');
    </script>
    <style>
        :root {
            --MAIN_TRANSCRIPT_BACKGROUND_COLOR: #eaf3ff;
            --CODING_TRANSCRIPT_BACKGROUND_COLOR: white;
            --NON_CODING_TRANSCRIPT_BACKGROUND_COLOR: #e5e5e5;
        }

        .coding-transcript {
            background-color: var(--CODING_TRANSCRIPT_BACKGROUND_COLOR);
        }

        .non-coding-transcript {
            background-color: var(--NON_CODING_TRANSCRIPT_BACKGROUND_COLOR);
        }

        .main-transcript {
            background-color: var(--MAIN_TRANSCRIPT_BACKGROUND_COLOR);
        }

        body {
            overflow-x: scroll;
        }

        body, .ui.form, label {
            font-size: 11pt !important;
        }

        td {
            padding: 5px 15px;
        }

        .form-checkbox {
            margin-left: 15px;
            margin-bottom: 10px;
        }

        .igv-track-checkbox {
            margin-right: 15px;
            margin-bottom: 10px;
        }

        #pangolin-header th {
            border-top: 1px solid rgba(34, 36, 38, .15);
        }

        table {
            font-size: 11pt !important;
            border-collapse: collapse !important;
            border-left: 0px !important;
            border-right: 0px !important;
            width: 100%;
        }

        .table-header {
            color: #666666 !important;
        }

        th div {
            font-weight: normal;
        }

        .small-link {
            font-size: 10pt;
            margin-left: 5px;
            margin-top: 3px;
            display: inline-block;
        }

        .only-large-screen {
            display: none !important;
        }

        .transcript-color-legend {
            border-style: solid;
            min-width: 23px;
            display: inline-block;
            margin-left: 20px;
        }

        @media screen and (min-width: 770px) {
            .only-large-screen {
                display: block !important;
            }
        }

        .results-div {
            width: 100%;
            position: static;
        }

        @media screen and (max-width: 1400px) {
            .results-div {
                width: 85vw;
                position: relative;
            }
        }

    </style>
</head>
<body>
<div class="ui stackable grid">
    <div class="row">
        <div class="one wide column only-large-screen"></div>
        <div class="ten wide column">
            <div class="ui grid">
                <div class="row">
                    <div class="twelve wide column">
                        <div style="padding:25px 0px">
                            <div>
                                <div style="display:inline-block; ">Enter a variant below to see its
                                    <a href="https://pubmed.ncbi.nlm.nih.gov/30661751" target="_blank">SpliceAI</a>
                                    and <a href="https://github.com/tkzeng/Pangolin" target="_blank">Pangolin</a> scores
                                </div>
                                <a id="more-details1-button" href="#" style="float:right; padding-right:7px" onclick="$('#more-details1').show();$('#more-details1-button').hide();">[more details]</a>
                                <div id="more-details1" style="display:none">
                                    <br />
                                    <a href="https://github.com/illumina/SpliceAI" target="_blank">SpliceAI</a> is a copyrighted work by Illumina and is provided under GPLv3 and CC BY NC 4.0 licenses for academic and non-commercial use.
                                    This web service is maintained by the <a href="https://the-tgg.org/" target="_blank">TGG</a> at the <a href="https://www.broadinstitute.org/" target="_blank">Broad Institute</a>.
                                    It runs the SpliceAI and Pangolin splicing prediction models on user-specified variants and then displays the scores.<br />
                                    <br />
                                    For more information about SpliceAI, see <a href="https://doi.org/10.1016/j.cell.2018.12.015" target="_blank">Jaganathan et al, Cell 2019</a>
                                    and this <a href="https://www.youtube.com/watch?v=oJvhj-tYbBI" target="_blank">recorded talk</a> by Kishore Jaganathan.<br />
                                    For more information about Pangolin, see <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02664-4" target="_blank">Zeng & Li 2022</a>.<br />
                                    <br />

                                    Under the hood, this service runs modified versions of SpliceAI and Pangolin which have added support for multi-nucleotide variants (MNVs), provide access to
                                    raw scores in addition to delta scores, and have other customizations:<br />
                                    <a href="https://github.com/bw2/SpliceAI">https://github.com/bw2/SpliceAI</a><br />
                                    <a href="https://github.com/bw2/Pangolin">https://github.com/bw2/Pangolin</a><br />
                                    <br />
                                    The visualizations of SpliceAI and Pangolin results are implemented as custom <a href="https://github.com/igvteam/igv.js">IGV.js</a> tracks at: <br />
                                    <a href="https://github.com/bw2/igv.js/pull/1">https://github.com/bw2/igv.js</a><br />
                                    <br />
                                    We no longer use the Illumina precomputed SpliceAI score tables since they are based on an old version of Gencode. <br />
                                    <br />
                                    <b>NOTE:</b> This service supports no more than a handful of queries per-user per-minute. If you need to batch-process many variants, please use the
                                    <a href="https://github.com/broadinstitute/SpliceAI-lookup/blob/master/README.md#local-install" target="_blank">source code</a>
                                    to set up your own local instance of the API server or just run the SpliceAI and Pangolin models directly.<br />
                                    <br />

                                </div>
                            </div>
                            <br/>
                            <div style="padding-left: 0px; color: gray">
                                <table>
                                    <thead style="text-align: left"><tr><th style="padding-bottom: 20px;">Examples &nbsp; (on hg38):</th><th></th></tr></thead>
                                    <tbody>
                                    <tr><td>chr8-140300616-T-G</td><td></td></tr>
                                    <tr><td>6 &nbsp; 31740453 &nbsp; G &nbsp; T</td><td></td></tr>
                                    <tr>
                                        <td style="padding-right: 0px">NM_001089.3(ABCA3):c.875A>T (p.Glu292Val)</td>
                                        <td style="padding-left: 0px; padding-right:7px;">
                                            <span class="only-large-screen">
                                                <a id="more-details2-button" href="#" style="float:right;"
                                                    onclick="$('.more-details2').show();$('#more-details2-button').hide();">[show more examples]</a>
                                            </span>
                                        </td>
                                    </tr>
                                    <tr style="display:none;" class="more-details2"><td>8:140300616 T&gt;G</td><td>'chr' <i>prefix is optional</i></td></tr>
                                    <tr style="display:none;" class="more-details2"><td>1:930130:C:G</td><td><i>position with overlapping genes</i></td></tr>
                                    <tr style="display:none;" class="more-details2"><td>1-1042601-A-AGAGAG</td><td><i>insertion of GAGAG</i></td></tr>
                                    <tr style="display:none;" class="more-details2"><td>1-1042466-GGGC-G</td><td><i>deletion of GGC</i></td></tr>
                                    <tr style="display:none;" class="more-details2"><td>NM_000249.4(MLH1):c.116G>A</td><td><i>another HGVS example</i></td></tr>
                                    </tbody>
                                </table>
                            </div>
                        </div>
                        <!--  <br />
                        <div style="font-size: large; color: darkred">
                          <br />
                          <b>The server is currently down for updates</b><br />
                          <br />
                          Please check back in ~20 minutes.
                        </div>
                        <br />
                        <br />
            -->
                        <div class="ui form" style="width:100%">
                            <div class="ui input" style="padding-bottom:30px; padding-right:8px; width:100%">
                                <input id="search-box" type="text" style="width: 100%; padding: 7px 10px" placeholder="Enter a variant...">
                            </div>
                            <div class="field" style="padding-right:10px; white-space:nowrap">
                                <span style="display: inline-block; white-space: nowrap; padding-right:35px;">
                                    <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Genome version:</b></span>
                                    <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                        <input type="radio" name="hg" value="37"/><label>hg19</label>
                                    </div>
                                    <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                        <input type="radio" name="hg" value="38" checked/><label>hg38</label>
                                    </div>
                                </span>
                                <span style="display: inline-block; white-space: nowrap;">
                                    <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Gencode:</b></span>
                                    <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                        <input type="radio" name="gencode-gene-set" value="basic" checked/><label>basic</label>
                                    </div>
                                    <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                        <input type="radio" name="gencode-gene-set" value="comprehensive" /><label>comprehensive</label>
                                    </div>
                                </span>
                                <i style="margin-left:10px" class="question circle outline icon" data-position="right center"
                                   data-content="The basic gene set starts with the most useful Gencode transcripts, prioritizing full-length protein coding transcripts over partial or non-coding ones. The comprehensive gene set starts with all Gencode transcripts. To improve efficiency, we then filter both gene sets, removing redundant transcripts which differ only in their internal exon boundaries but share the same transcript strand, start, and end coordinates. This filtering approach is based on the fact that SpliceAI (unmasked) scores depend solely on pre-mRNA reference and alternate sequence (defined by transcript strand, start, and end coordinates) and not on exon boundaries."></i>
                            </div>
                            <div class="ui stackable grid">
                                <div class="row">
                                    <div class="thirteen wide column">
                                        <div class="field">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Max distance:</b></span>
                                            <div class="ui input" style="padding-left:15px; padding-bottom:10px; padding-right:8px; width:100px">
                                                <input id="max-distance-input" type="text" value="500" style="padding:7px 10px">
                                            </div>
                                            <i class="question circle outline icon" data-position="right center"
                                               data-content="For each variant, SpliceAI looks within a window (+/- 500bp by default) to see how the variant affects the probabilities of different positions being splice acceptors or donors. The distance specified here controls the size of this window. The maximum allowed value is 10,000bp."></i>

                                            <div class="ui checkbox form-checkbox"><input type="checkbox" name="mask"><label>masked scores</label></div>
                                            <i style="margin-left:10px" class="question circle outline icon" data-position="right center"
                                               data-content="Selecting this checkbox will mask scores that strengthen annotated splice sites or weaken unannotated splice sites and show 0 instead. Such changes are considered less likely to be pathogenic than those that weaken annotated splice sites and strengthen unannotated splice sites. Uncheck this box to show all scores. SpliceAI developers recommend using 'masked' scores for variant interpretation while unmasked scores are recommended for alternative splicing analysis."></i>

                                            <div class="ui checkbox form-checkbox"><input type="checkbox" name="show-ref-alt"><label>REF & ALT scores</label></div>
                                            <i style="margin-left:10px" class="question circle outline icon" data-position="right center"
                                               data-content="Selecting this checkbox will add two extra columns to the results. These will contain the raw SpliceAI and Pangolin predictions for the reference haplotype and the alternate haplotype. The delta scores represent the difference between these reference and alternate scores."></i>
                                        </div>
                                    </div>
                                    <div class="three wide column">
                                        <button id="submit-button" class="ui primary button" style="margin-bottom: 15px; margin-right:10px; float: right">Submit</button>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        </div>
        <div class="four wide column">
            <div class="only-large-screen" style="padding:25px 0px">
                <div>
                    <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues" target="_blank">SpliceAI-lookup/issues</a>: issues or feature requests for this website<br />
                    <br/>
                    <b>November 3, 2024</b><br />
                    - added choice of Gencode <a href="https://genome.ucsc.edu/FAQ/FAQgenes.html#gencode" target="_blank">basic or comprehensive</a> transcripts<br />
                    - updated to Gencode v47 <br />
                    <br/>
                    <b>June 20, 2024</b><br />
                    - fixed SpliceAI visualizations to clarify positions of predicted splicing changes. See <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues/70" target="_blank">issue #70</a> for details. <br />
                    <br />
                    <a id="more-details3-button" href="#"
                       onclick="$('#more-details3').show();$('#more-details3-button').hide();">[show older updates]</a>
                    <div id="more-details3" style="display:none">
                        <b>June 3, 2024</b><br />
                        - show the input variant's <a href="https://useast.ensembl.org/info/docs/tools/vep/index.html" target="blank">VEP</a> consequence in the results table<br />
                        - the Gencode gene track is now optional in the visualizations<br />
                        - updated everything to <a href="https://www.gencodegenes.org/human/releases.html">Gencode v46</a><br/>
                        <br />
                        <b>March 7, 2024</b><br/>
                        - added warning for insertion variants with delta scores ≥ 0.2 and position = 0bp saying that they may be difficult to interpret due to <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues/69">issue #67</a>. Thanks to @SophieCandille for the issue report and example variant: <a href="https://spliceailookup.broadinstitute.org/#variant=NM_000179.3%3Ac.261-3_261-2insAGTTG&hg=38&distance=500&mask=0&ra=1" target="_blank">2:47790924 C>CAGTTG</a><br />
                        - moved server to Google Cloud Run to better support higher usage<br />
                        <br />
                        <b>November 30, 2023</b><br/>
                        - changed default search settings to "not masked" due to <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues/59#issuecomment-1815667021">confusion</a> and <a href="https://github.com/Illumina/SpliceAI/issues/27#issuecomment-547089167">issues</a> with how masked scores are computed. <br />
                        - when looking at hg19 variants, the visualization checkboxes now make it clear which IGV.js tracks are or aren't available on hg19.<br/>
                        <br/>
                        <b>October 5, 2023</b><br/>
                        - fixed Pangolin bug identified by <a href="https://github.com/invitae/pangolin/pull/5">@kazmiekr</a> where variants that overlap more than one gene showed incorrect masked scores<br />
                        - <a href="https://github.com/pj-sullivan">@pj-sullivan</a> added controls to show/hide transcripts<br/>
                        - added MANE Select transcript labels<br/>
                        - updated to gencode v44 "basic" annotations (instead of comprehensive). These have fewer transcripts per gene.<br/>
                        - added igv.js visualizations<br/>
                        <br/>
                        <b>May 22, 2023</b><br/>
                        - changed defaults to 'masked' and max distance to 500bp<br/>
                        - retired <b>Illumina precomputed scores</b> since they were created using Gencode v24
                        and max distance = 50bp and so can differ from computed scores which are based on the latest Gencode.<br/>
                        <br/>
                        <b>May 12, 2023</b><br/>
                        - added REF & ALT score columns to show SpliceAI's underlying predictions for each haplotype
                        (using <a href="https://github.com/Illumina/SpliceAI/pull/92">SpliceAI PR by @h-joshi</a>).
                        The &#916;&nbsp; score column is the difference between these two scores.<br/>
                        - added support for MNPs (see <a
                            href="https://github.com/broadinstitute/SpliceAI-lookup/issues/1">issue #1</a> & <a
                            href="https://github.com/Illumina/SpliceAI/pull/119">SpliceAI PR by @kdahlo</a>)<br/>
                        - updated to <a href="https://www.gencodegenes.org/human/releases.html">Gencode v43</a> (was previously on v42)<br/>
                        <br/>
                        <b>May 10, 2023</b><br/>
                        - fixed bug where "raw" Pangolin scores were shown regardless of whether the user selected "raw" or "masked".<br/>
                        <br/>

                        <b>Feb 5, 2023</b><br/>
                        - show <a href="https://github.com/tkzeng/Pangolin">Pangolin scores</a> in addition to SpliceAI scores.<br/>
                        <br/>
                        <b>June 2, 2021</b><br/>
                        - show RefSeq ids for the subset of Ensembl ENST ids (~30%) that have one or more matching
                        RefSeq NM ids according to Ensembl. See <a
                            href="https://github.com/broadinstitute/SpliceAI-lookup/issues/8">issue #8</a> for details.<br/>
                        <br/>
                        <b>April 11, 2021</b><br/>
                        - show gray background for non-coding transcripts<br/>
                        - switch to showing all non-coding transcripts. (Previously, transcripts with Gencode biotypes
                        like <i>lncRNA</i>, <i>processed_pseudogene</i>, <i>processed_transcript</i>,
                        <i>retained_intron</i>, and
                        <i>nonsense_mediated_decay</i> were filtered out if they overlapped <i>protein_coding</i> transcripts).
                        See <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues/6#issuecomment-816942824">issue #6</a> for details.<br/>
                        <br />
                        <a id="hide-details3-button" href="#" onclick="$('#more-details3').hide();$('#more-details3-button').show();">[hide older updates]</a>
                    </div>
                    <br />
                    <br/>
                    <div>
                      <b>Related web tools:</b><br/>
                        <a href="https://liftover.broadinstitute.org" target="_blank">liftover</a>: for variants/positions/intervals (hg19 <=> hg38 <=> T2T)<br/>		      
                <a href="https://tgg-viewer.broadinstitute.org" target="_blank">TGG Viewer</a>: igv.js-based web viewer for public reference tracks and private data in Google Storage buckets. Has custom track types for RNA-seq <a href="https://github.com/igvteam/igv.js/pull/1019#issue-521944897" target="_blank">splice junctions</a> and <a href="https://github.com/igvteam/igv.js/pull/1055#issue-560042328" target="_blank">gCNV</a> variants.<br/>
                    </div>
                </div>
            </div>
        </div>
        <div class="one wide column only-large-screen"></div>
    </div>
    <div class="row">
        <div class="one wide column only-large-screen"></div>
        <div class="fourteen wide column">
            <div class="results-div">
                <div id="response-box" style="display: none">
                    <table id="spliceai-table" class="ui stackable table">
                        <thead id="spliceai-header">
                        <tr>
                            <th style="background-color:white" colspan="7">SpliceAI scores: 
                                <i class="question circle outline icon" data-position='right center'
                                   data-content="SpliceAI scores are computed by running the SpliceAI model with user-specified parameters (ie. max distance)."></i>
                            </th>
                        </tr>
                        <tr>
                            <th class="table-header">Variant</th>
                            <th class="table-header" style="white-space: nowrap">Gene
                                <div style="float:right;">
                                    <span class='transcript-color-legend'
                                          style="background-color: var(--MAIN_TRANSCRIPT_BACKGROUND_COLOR)!important">&nbsp;</span>
                                    = <span class="main-transcript-label">MANE Select</span> transcript
                                    <span class='transcript-color-legend'
                                          style="background-color: var(--NON_CODING_TRANSCRIPT_BACKGROUND_COLOR)!important">&nbsp;</span>
                                    = non-coding transcript
                                </div>
                            </th>
                            <th class="table-header">&#916;&nbsp; type</th>
                            <th class="table-header" style="white-space: nowrap">&#916;&nbsp; score &nbsp; <i class='question circle outline icon'
                                                                                 data-html="Delta scores range from 0 to 1 and can be interpreted as the probability that the variant affects splicing at any position within a window around it (+/- 500bp by default). In the SpliceAI paper, a detailed characterization is provided for 0.2 (high recall), 0.5 (recommended), and 0.8 (high precision) cutoffs. Consequently, scores of 0.2 or higher appear in <b style='color:#1fb839'>green</b>, 0.5 or higher appear in <b style='color:#FFCB1F'>yellow</b>, and 0.8 or higher appear in <b style='color:#FF0000'>red</b>."></i>
                            </th>
                            <th class="table-header" style="white-space: nowrap">position &nbsp; <i class='question circle outline icon'
                                                                                data-content="For each variant, SpliceAI looks within a window (+/- 500bp by default) to see how the variant affects the probabily of different positions serving as splice acceptors or donors. The offsets in this column represent positions with the biggest change in probability within the window. Negative values are to the left of the variant in genomic coords, regardless of transcript strand."></i>
                            </th>
                            <th class="table-header ref-score-column" style="white-space: nowrap">
                                REF score &nbsp; <i class='question circle outline icon'
                                            data-content="SpliceAI's computed probability that the given position is a splice acceptor or donor based on the reference haplotype sequence. SpliceAI computes delta scores by taking the difference between the REF and ALT score at each position."></i>
                            </th>
                            <th class="table-header alt-score-column" style="white-space: nowrap">
                                ALT score &nbsp; <i class='question circle outline icon'
                                            data-content="SpliceAI's computed probability that the given position is a splice acceptor or donor based on the alternate haplotype sequence. SpliceAI computes delta scores by taking the difference between the REF and ALT score at each position."></i>
                            </th>
                        </tr>
                        </thead>
                    </table>
                    <table id="transcript-button-table" class="ui stackable table" style="border: none !important">
                        <tbody>
                        <tr>
                            <td style="padding: 10px 0px 10px 0px">
                                <div class="ui buttons">
                                    <button id="main-transcript-button" class="ui button" onclick="updateTranscriptButtons('main')">
                                        <span class="main-transcript-label">MANE Select</span> Transcript
                                    </button>
                                    <div class="or"></div>
                                    <button id="all-transcript-button" class="ui button" onclick="updateTranscriptButtons('all')">
                                        All Transcripts
                                    </button>
                                </div>
                            </td>
                        </tr>
                        </tbody>
                    </table>
                    <table id="pangolin-table" class="ui stackable table">
                        <thead id="pangolin-header">
                            <tr>
                                <th style="background-color:white" colspan="7">Pangolin scores: <i class='question circle outline icon' data-position='right center'
                                                           data-content="Pangolin scores are computed by running the Pangolin model with user-specified parameters (ie. max distance)."></i>
                                </th>
                            </tr>
                            <tr>
                                <th class="table-header">Variant</th>
                                <th class="table-header">Gene</th>
                                <th class="table-header">&#916;&nbsp; type</th>
                                <th class="table-header" style="white-space: nowrap">&#916;&nbsp; score &nbsp; <i class='question circle outline icon'
                                                                                     data-html="Delta scores range from 0 to 1 and can be interpreted as the probability that the variant affects splicing at any position within a window around it (+/- 500bp by default). Pangolin does not recommend specific score thresholds; however, here scores of 0.2 or higher appear in <b style='color:#1fb839'>green</b>, 0.5 or higher appear in <b style='color:#FFCB1F'>yellow</b>, and 0.8 or higher appear in <b style='color:#FF0000'>red</b>."></i>
                                </th>
                                <th class="table-header" style="white-space: nowrap">position &nbsp; <i class='question circle outline icon'
                                                                                    data-content="For each variant, Pangolin looks within a window (+/- 500bp by default) to see how the variant affects the probability of splicing occuring at different positions. The offsets in this column represent positions with the biggest change in probability within the window. Negative values are upstream (5&#039;) of the variant, and positive are downstream (3&#039;) of the variant."></i>
                                </th>
                                <th class="table-header ref-score-column" style="white-space: nowrap">
                                    REF score &nbsp; <i class='question circle outline icon'
                                    data-content="Pangolin's computed probability that the given position is a splice site based on the reference haplotype sequence."></i></th>
                                <th class="table-header alt-score-column" style="white-space: nowrap">
                                    ALT score &nbsp; <i class='question circle outline icon'
                                                data-content="Pangolin's computed probability that the given position is a splice site based on the alternate haplotype sequence."></i></th>
                            </tr>
                        </thead>
                    </table>
                    <table id="igv-table" class="ui stackable table" style="margin-top: 30px; display: none">
                        <thead style="background-color:white">
                        <tr>
                            <th style="background-color:white" colspan="5"><b>Visualizations:</b></th>
                        </tr>
                        <tr>
                            <th class="table-header"><span style="white-space: nowrap">SpliceAI &amp; Pangolin results
                                <i class="question circle outline icon" data-position="right center"
                                   data-content="These tracks display the computed SpliceAI and Pangolin scores for the variant entered by the user."></i>
                                </span>
                            </th>
                            <th class="table-header"><span style="white-space: nowrap">GTEx RNA: average per sample
                                <i class="question circle outline icon" data-position="right center"
                                   data-content="These tracks summarize all splice junctions and per-base read depths in GTEx v8 RNA-seq samples for a specific tissue. Both the read depths and splice junction read counts are divided by the number of samples, and so represent the per-sample average. These tracks are only available for hg38."></i>
                                </span>
                            </th>
                            <th class="table-header"><span style="white-space: nowrap">GTEx RNA: all splice junctions
                                <i class="question circle outline icon" data-position="right center"
                                   data-content="These tracks show the splice junctions and per-base read depths in GTEx v8 RNA-seq samples from a specific tissue. The read depths and splice junction read counts are summed across all samples from that tissue. These tracks are only available for hg38."></i>
                                </span>
                            </th>
                            <th class="table-header"><span style="white-space: nowrap">SpliceAI: precomputed scores
                                <i class="question circle outline icon" data-position="right center"
                                   data-content="These tracks summarize the original precomputed SpliceAI score tables made available by Illumina. They are only available for hg38."></i>
                                </span>
                            </th>
                            <th class="table-header"><span style="white-space: nowrap">Reference
                                <i class="question circle outline icon" data-position="right center"
                                   data-content="Miscellaneous reference tracks."></i>
                                </span>
                            </th>
                        </tr>
                        </thead>
                        <tr>
                            <td>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-variant" checked><label>variant track</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-spliceai-ref-alt" checked><label>SpliceAI REF &amp; ALT scores</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-spliceai-delta-scores" checked><label>SpliceAI &#916;&nbsp; scores</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-pangolin-delta-scores" checked><label>Pangolin &#916;&nbsp; scores</label></div>
                            </td>
                            <td>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-gtex-blood"><label>blood</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-gtex-fibroblasts"><label>fibroblasts</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-gtex-muscle"><label>muscle</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-gtex-lymphocytes"><label>LCLs</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-gtex-brain-cortex"><label>brain cortex</label></div>
                            </td>
                            <td>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-gtex-blood-all"><label>blood</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-gtex-fibroblasts-all"><label>fibroblasts</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-gtex-muscle-all"><label>muscle</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-gtex-lymphocytes-all"><label>LCLs</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-gtex-brain-cortex-all"><label>brain cortex</label></div>
                            </td>
                            <td>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-spliceai-precomputed-gain-0.5"><label>gain ≥ 0.5</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-spliceai-precomputed-loss-0.5"><label>loss ≥ 0.5</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-spliceai-precomputed-gain-0.2"><label>gain ≥ 0.2</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-spliceai-precomputed-loss-0.2"><label>loss ≥ 0.2</label></div>
                            </td>
                            <td>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-gencode-genes"><label>gencode <span id="gencode-version"></span> genes</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-100-mer-mappability"><label>100-mer mappability</label></div>
                                <div class="ui checkbox igv-track-checkbox"><input type="checkbox" name="igv-segdups"><label>segmental duplications</label></div>
                            </td>
                        </tr>
                        <tr>
                            <td colspan="5" style="padding-left: 0px">
                                <button id="show-igv-button" class="ui primary button" style="margin-bottom: 15px">Show</button>
                            </td>
                        </tr>
                    </table>
                    <div id="igv-div" style="width:100%; padding-top:5px; display: none">
                        <div style="padding:20px 0px; font-size:110%">To see <b>documentation</b> or <b>legends</b> for each track, click on the rectangle that contains the track name.</div>
                        <div id="igv-viewport" style="width:100%; min-height:500px"></div>
                    </div>
                </div>
                <div id="error-box" style="color:darkred; padding:10px 0px; font-size:medium"></div>
            </div>
        </div>
        <div class="one wide column only-large-screen"></div>
    </div>
</div>

<script>
    const GENCODE_VERSION = "v47"

    const VARIANT_RE = new RegExp(
        "^[\\s]*" +
        "(chr)?([0-9XYMTt]{1,2})" +
        "[-\\p{Pd}\\s:]+" +
        "([0-9,]+)" +
        "[-\\p{Pd}\\s:]*" +
        "([ACGT]+)" +
        "[-\\p{Pd}\\s:>]+" +
        "([ACGT]+)",
        'iu'
    )

    const TRANSCRIPT_PRIORITY = {
        "MS": 3,
        "MP": 2,
        "C": 1,
        "N": 0,
    }

    // store user settings at the time of the last 'submit', as well as the resulting responses from the server
    let lastGenomeVersion, lastSpliceaiResponseJson, lastPangolinResponseJson

    const formatScore = (score) => {
        if (score == null) {
            return ""
        }

        return Math.abs(parseFloat(score)).toFixed(2)
    }

    const getScoreStyle = (score) => {
        score = parseFloat(score)

        if (Math.abs(score) < 0.01) {
            return `style='color:#BBBBBB;'`
        }

        let color
        if (Math.abs(score) >= 0.8) {
            color = "#fccfb8"
        } else if (Math.abs(score) >= 0.5) {
            color = "#fff19d"
        } else if (Math.abs(score) >= 0.2) {
            color = "#cdffd7"
        } else {
            return ""
        }

        return `style='white-space:nowrap;background-color:${color};'`
    }

    const { TabixIndexedFile } = window.gmodTABIX

    const lookupTableUrls = {
        '37': 'https://storage.googleapis.com/spliceai-lookup-reference-data/PrimateAI_3D.hg19.with_gene_thresholds.txt.gz',
        '38': 'https://storage.googleapis.com/spliceai-lookup-reference-data/PrimateAI_and_PromoterAI_scores.hg38.tsv.gz',
    }

    const lookupTables = {
        '37': new TabixIndexedFile({ url: lookupTableUrls['37'], tbiUrl: `${lookupTableUrls['37']}.tbi` }),
        '38': new TabixIndexedFile({ url: lookupTableUrls['38'], tbiUrl: `${lookupTableUrls['38']}.tbi` }),
    }

    const getUCSCBrowserUrl = (genomeVersion, chrom, pos) => {
        const genomeVers = genomeVersion.replace('37', '19')
        const chromWithoutPrefix = chrom.toUpperCase().replace('CHR', '')

        return `https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg${genomeVers}&position=chr${chromWithoutPrefix}:${pos}`
    }

    const showError = (errorMessage) => {
        $("#error-box").html( $("#error-box").html() + `<br />${errorMessage.toString()}<br />`)
        $("#error-box").show()
    }

    const normalizeVariant = async (variant, genomeVersion) => {
        /* Convert the given variant to a standardized "{chrom}-{pos}-{ref}-{alt}" string using a regular expression, or
         * if that fails, assume the variant is in HGVS notation and try using the Ensembl hgvs API to convert it.
         *
         * Args:
         *  variant (string): user input text
         *  genomeVersion (string): "37" or "38"
         *
         * Return:
         *  string: the input variant reformatted as a "{chrom}-{pos}-{ref}-{alt}" string
         */

        const ensemblApiPrefix = `https://${genomeVersion == '37' ? 'grch37.' : ''}rest.ensembl.org/vep/human/hgvs/`

        let chrom, pos, ref, alt, variantConsequence
        let matchedRegExp = variant.match(VARIANT_RE)
        let ensemblApiUrl, ensemblApiResponse, ensemblApiResponseJson
        if (matchedRegExp) {
            // was able to parse the user input using a simple reg-exp
            chrom = matchedRegExp[2].toUpperCase()
            pos = parseInt(matchedRegExp[3].replace(/,/g, ""))
            ref = matchedRegExp[4].toUpperCase()
            alt = matchedRegExp[5].toUpperCase()

            if (ref.length == 1) {
                //handle SNPs and insertions
                ensemblApiUrl = `${ensemblApiPrefix}${chrom}:g.${pos}${ref}>${alt}`
            } else if(alt.length == 1) {
                //handle deletions
                ensemblApiUrl = `${ensemblApiPrefix}${chrom}:g.${pos+1}_${pos+ref.length-1}del${ref.slice(1)}`
            } else {
                //converting MNP into HGVS is not straight forward, so just return the variant
                return { 'variant': `${chrom}-${pos}-${ref}-${alt}` }
            }
        }
        else {
            //assume the variant is already in HGVS notation
            ensemblApiUrl = `${ensemblApiPrefix}${variant.trim()}`
        }


        // try calling the Ensembl API on the user input
        try {
            try {
                ensemblApiResponse = await makeRequest(ensemblApiUrl + "?content-type=application/json&vcf_string=1")
            } catch (e) {
                console.error(e)
                throw Error(`Ensembl API call failed: Unable to reach server`)
            }

            ensemblApiResponseJson = await ensemblApiResponse
            console.log(`Ensembl API response:`, ensemblApiResponseJson)

            if (!ensemblApiResponse.ok || ensemblApiResponseJson.error) {
                let errorText = `${ensemblApiResponseJson.error}`
                //const unableToParseHGVSErrorMatch = errorText.match("Unable to parse HGVS notation")
                //if (unableToParseHGVSErrorMatch) {
                //    errorText = `Ensembl API is unable to parse the variant ${variant}: ${errorText}`
                //}
                const refAlleleErrorMatch = errorText.match(
                    new RegExp("[(]([ACGTRYSWKMBDHVN]+)[)] does not match reference allele given by HGVS notation"))
                if (refAlleleErrorMatch) {
                    errorText = `${variant} has an unexpected reference allele. The hg${genomeVersion} reference allele should be ${refAlleleErrorMatch[1]}`
                }

                throw new Error(errorText);
            }

            if (!ensemblApiResponseJson[0] || !ensemblApiResponseJson[0].vcf_string) {
                throw new Error(`Unexpected response: ${ensemblApiResponseJson}`);
            }

            variant = ensemblApiResponseJson[0].vcf_string
            variantConsequence = ensemblApiResponseJson[0].most_severe_consequence
            matchedRegExp = variant.match(VARIANT_RE)
            if (!matchedRegExp) {
                throw new Error(`Unexpected response: ${ensemblApiResponseJson}`)
            }

            chrom = matchedRegExp[2].toUpperCase()
            pos = parseInt(matchedRegExp[3])
            ref = matchedRegExp[4].toUpperCase()
            alt = matchedRegExp[5].toUpperCase()

            result = {
                'variant': `${chrom}-${pos}-${ref}-${alt}`,
                'consequence': variantConsequence,
            }
            console.log("EnsemblAPI result:", result)

            return result

        } catch (e) {
            console.error(e)
            throw new Error(e)
        }



        /*
        // left-align the indel
        let leftAlignApiResponse
        try {
            leftAlignApiResponse = await makeRequest(`${baseApiUrl['normalize']}/normalize/?g=hg${genomeVersion}&chrom=${chrom}&pos=${pos}&ref=${ref}&alt=${alt}`)
            const leftAlignApiResponseJson = await leftAlignApiResponse
            console.log(`Left-align API response:`, leftAlignApiResponseJson)
            if (!leftAlignApiResponse.ok || leftAlignApiResponseJson.error) {
                console.log(`${leftAlignApiResponseJson.error}`)
            } else {
                if (leftAlignApiResponseJson.normalized_ref && leftAlignApiResponseJson.normalized_alt && leftAlignApiResponseJson.normalized_ref.length != leftAlignApiResponseJson.normalized_alt.length && (
                    parseInt(leftAlignApiResponseJson.normalized_pos) != pos || leftAlignApiResponseJson.normalized_ref != ref || leftAlignApiResponseJson.normalized_alt != alt)
                ) {
                    chrom = leftAlignApiResponseJson.normalized_chrom
                    pos = leftAlignApiResponseJson.normalized_pos
                    ref = leftAlignApiResponseJson.normalized_ref
                    alt = leftAlignApiResponseJson.normalized_alt
                }
            }

        } catch(e) {
            console.error(e)
        }

        return `${chrom}-${pos}-${ref}-${alt}`

        */
    }

    const makeRequest = (url) => {
        const method = "GET"
        return new Promise(async (resolve, reject) => {
            const xhr = new XMLHttpRequest()
            xhr.open(method, url)
            xhr.addEventListener("load", () => {
                console.log("GET", url, xhr.status, xhr.statusText)
                let response = {}
                try {
                    response = JSON.parse(xhr.response)
                } catch(e) {
                    console.error("Unable to parse response", xhr.response)
                    reject(`Unexpected error: ${url}`)
                    return
                }

                response.ok = xhr.status >= 200 && xhr.status < 300
                resolve(response)
            })
            xhr.addEventListener("error", () => {
                if (xhr.status == 0) {
                    reject("Unable to reach server")
                } else {
                    reject(`${xhr.status}  ${xhr.statusText}`)
                }
            })
            //console.log('Sending request', method, url)
            xhr.send()
        })
    }

    const considerInsertedBases = (score, position, ref, alt, scoresForInsertedBases) => {
        //check if the variant is an insertion and if so, return the number of inserted bases
        return score >= 0.01 && position == 0 && ref.length == 1 && alt.length > 1 && scoresForInsertedBases && scoresForInsertedBases.length > 0
    }

    const generateTableOfScoresForInsertedBases = (modalId, score, position, ref, alt, scoresForInsertedBases) => {
        //warn about insertion variants that are predicted to cause a donor or acceptor gain somewhere within the inserted sequence.
        if (!considerInsertedBases(score, position, ref, alt, scoresForInsertedBases)) {
            return ""
        }

        let tableRows = []
        for (const scoreObj of scoresForInsertedBases) {
            const RA = parseFloat(scoreObj.RA)
            const RD = parseFloat(scoreObj.RD)
            const AA = parseFloat(scoreObj.AA)
            const AD = parseFloat(scoreObj.AD)

            tableRows.push(
                `<tr><td>${scoreObj.chrom}</td>
                     <td style='text-align:right'>${scoreObj.pos}</td>
                     <td>${scoreObj.ref}</td>
                     <td>${scoreObj.alt}</td>
                     <td ${getScoreStyle(RA)}>${RA}</td>
                     <td ${getScoreStyle(RD)}>${RD}</td>
                     <td ${getScoreStyle(AA)}>${AA}</td>
                     <td ${getScoreStyle(AD)}>${AD}</td></tr>`)
        }

        const table = (
            `<div>
               Detailed SpliceAI predictions for bases within and around the inserted sequence: <br>
               <table class='ui celled table'>
                <thead>
                    <tr style='text-align:center'>
                        <th style='width:1%'>chrom</th>
                        <th>position</th>
                        <th style='width:1%'>REF</th>
                        <th style='width:1%'>ALT</th>
                        <th style='width:1%'>REF acceptor score</th>
                        <th style='width:1%'>REF donor score</th>
                        <th style='width:1%'>ALT acceptor score</th>
                        <th style='width:1%'>ALT donor score</th>
                    </tr>
                </thead>
                <tbody>
                    ${tableRows.join("\n")}
                </tbody>
            </table></div>`
        ).replace(/\n/g, "")

        //check if browser is chrome, safari or firefox
        //const isChrome = /Chrome/.test(navigator.userAgent) && /Google Inc/.test(navigator.vendor)
        //const isSafari = /Safari/.test(navigator.userAgent) && /Apple Computer/.test(navigator.vendor)
        //const isFirefox = /Firefox/.test(navigator.userAgent)

        const popupIcon = `<i style="margin-left:10px; color:#000000" class="table icon score-table" data-position="right center" data-html="${table}"></i>`
        const modal = `<div id='modal-with-table${modalId}' class='ui modal'><i class='close icon'></i><div class='scrolling content'>${table}</div></div>`
        const modalIcon = `<i id='table-icon${modalId}' style='margin-left:10px;color:#000080;cursor:pointer' class='window maximize outline icon'></i>`
        return `${modalIcon}${modal} ${popupIcon}`

    }

    const updatePositionAccountingForInsertedBases = (scoreKey, score, position, ref, alt, scoresForInsertedBases) => {
        if (!considerInsertedBases(score, position, ref, alt, scoresForInsertedBases)) {
            return `${position} bp`
        }

        if (scoreKey == 'DS_AG') {
            scoreKey = 'AA'
        } else if (scoreKey == 'DS_DG') {
            scoreKey = 'AD'
        } else {
            return `${position} bp`
        }
        // find the position within scoresForInsertedBases at which the score is the highest. Look at the AA score if scoreKey == 'DS_AG', and AD score if scoreKey == 'DS_DG'
        let maxScore = -1
        let maxScorePosition = 0
        for (const scoreObj of scoresForInsertedBases) {
            const scoreValue = parseFloat(scoreObj[scoreKey])
            if (scoreValue > maxScore) {
                maxScore = scoreValue
                maxScorePosition = parseInt(scoreObj.pos)
            }
        }

        return `+${maxScorePosition} bp position within the inserted sequence`
    }

    const getGnomadDataVersion = (genomeVersion) => {
        return genomeVersion === "38" ? "gnomad_r4" : "gnomad_r2_1"
    }

    const generateSplicingResultsTable = async (normalizedVariant, variantConsequence, tool, variant, genomeVersion, basicOrComprehensive, maxDistance, mask, showRefAltScoreColumns) => {
        /* Generate the results table to show either the SpliceAI or Pangolin scores */

        const variantTokens = (normalizedVariant || "---").split("-")
        const chrom = variantTokens[0]
        const pos = variantTokens[1]
        const ref = variantTokens[2]
        const alt = variantTokens[3]

        const baseUrl = baseApiUrl[`${tool.toLowerCase()}-${genomeVersion}`]
        const urlArgs = `hg=${genomeVersion}&bc=${basicOrComprehensive}&distance=${maxDistance}&mask=${mask}&variant=${chrom}-${pos}-${ref}-${alt}&raw=${variant}&variant_consequence=${variantConsequence}`

        let apiResponse
        try {
            apiResponse = await makeRequest(`${baseUrl}/${tool.toLowerCase()}/?${urlArgs}`)
        } catch(e) {
            throw Error(`${tool} API call failed: ${e}`)
        }

        const apiResponseJson = await apiResponse
        console.log(`${tool} API response:`, apiResponseJson)

        if (!apiResponse.ok || apiResponseJson.error) {
            throw Error(`${tool} API call error ${apiResponseJson.error ? `: ${apiResponseJson.error}` : ""}`)
        }

        variantConsequence = variantConsequence || ""
        const variantConsequenceDiv = variantConsequence ? `
            <div display: inline-block>
                <a href="https://www.ensembl.org/info/genome/variation/prediction/predicted_data.html" class="small-link" target="_blank">
                ${variantConsequence.replace(/_/g, ' ')}
                </a>
             </div>
             <br class="only-large-screen"/>
             ` : ""

        //sort transcripts
        apiResponseJson.scores = _(apiResponseJson.scores).sortBy((s) => (  //compute sort key
                100*(s['t_priority'].startsWith("M") ? 0 : 1) +
                10* (s['t_type'] == "protein_coding" ? 0 : 1) +
                -1*  TRANSCRIPT_PRIORITY[s['t_priority']]
        )).values()


        const TRANSCRIPT_PRIORITY_VIEW_LOOKUP = {
            "MS": `<a href="https://www.ncbi.nlm.nih.gov/refseq/MANE" target="_blank">MANE Select transcript</a>`,
            "MP": `<a href="https://www.ncbi.nlm.nih.gov/refseq/MANE" target="_blank">MANE Plus Clinical transcript</a>`,
            "C": "Canonical transcript",
        }

        const noDifferenceDueToNormalization = variant.toLowerCase().trim().replace(/^chr/, "").replace(/[>: _-]+/g, "-") == normalizedVariant.toLowerCase().trim().replace(/^chr/, "").replace(/[>: _-]+/g, "-")
        const normalizedVariantDiv = noDifferenceDueToNormalization ? "" :
            `<br class="only-large-screen"/>
             <div style="margin-left:5px;margin-right:10px; display: inline-block; color:#333333">
                <i>⇒ ${chrom}:${pos} ${ref}&gt;${alt}</i>
             </div>
             <br class="only-large-screen"/>`


        const transcriptCategories = {}
        const tableRows = []
        const modalDialogIds = []
        let transcriptIndex = 0
        for (const scores of apiResponseJson.scores) {
            const isPangolin = tool.toLowerCase() == "pangolin"
            const subRowCount = isPangolin ? 2 : 4
            const strand = scores['t_strand'] == "-" ? "minus" : "plus"
            const gnomadVersion = getGnomadDataVersion(genomeVersion)
            const isMainTranscript = scores['t_priority'] != "N" && (scores['t_priority'] != "C" || transcriptCategories["MS"] == undefined)
            const refSeqLink = scores['t_refseq_ids']? `/ <a href="https://www.ncbi.nlm.nih.gov/search/all/?term=${scores['t_refseq_ids'][0]}" target="_blank">${scores['t_refseq_ids'][0]}</a>` : ""

            transcriptCategories[scores['t_priority']] = true

            const resultRowClasses = [`${tool.toLowerCase()}-result-row`]
            if (isMainTranscript) {
                resultRowClasses.push("main-transcript")
            } else {
                resultRowClasses.push("non-main-transcript")
            }
            if(scores['t_type'] == "protein_coding") {
                resultRowClasses.push("coding-transcript")
            } else {
                resultRowClasses.push("non-coding-transcript")
            }

            $(`#${tool.toLowerCase()}-header`).nextAll().remove()
            let row = `<tr class="${resultRowClasses.join(' ')}">
                <td rowspan="${subRowCount}" style="vertical-align: top">
                    <div style="margin-right:10px; display: inline-block">${variant}</div><br class="only-large-screen"/>
                    ${normalizedVariantDiv}
                    <br class="only-large-screen"/>
                    ${variantConsequenceDiv}
                    <a href="${getUCSCBrowserUrl(genomeVersion, chrom, pos)}" class="small-link" target="_blank">UCSC</a>,
                    <a href="https://gnomad.broadinstitute.org/variant/${normalizedVariant}?dataset=${gnomadVersion}" class="small-link" target="_blank">gnomAD</a>
                    <br class="only-large-screen"/>

                </td>
                <td rowspan="${subRowCount}" style="vertical-align: top">
                    <div style="margin-right:10px; display: inline-block">
                        ${scores['g_name']}
                            <div class="small-link"> (
                                <a href="https://useast.ensembl.org/Homo_sapiens/Gene/Summary?g=${scores['g_id'].split('.')[0]}" target="_blank">${scores['g_id']}</a>
                                 / <a href="https://useast.ensembl.org/Homo_sapiens/Transcript/Summary?t=${scores['t_id'].split('.')[0]}" target="_blank">${scores['t_id']}</a>
                                 ${refSeqLink})
                            </div><br />
                            <br class="only-large-screen" />
                        <div class="small-link"><a href="https://www.gencodegenes.org/pages/biotypes.html" target="_blank">${scores['t_type'].replace(/_/g, " ")}</a></div>
                        <div class="small-link">${scores['t_priority'] != "N" ? TRANSCRIPT_PRIORITY_VIEW_LOOKUP[scores['t_priority']] : ""}</div>
                        <div class="small-link"> (${strand} strand)</div>
                    </div><br class="only-large-screen" />
                    <br class="only-large-screen" />
                    <a href="https://www.omim.org/search?search=${scores['g_name']}" class="small-link" target="_blank">OMIM</a>,
                    <a href="https://gtexportal.org/home/gene/${scores['g_name']}" class="small-link" target="_blank">GTEx</a>,
                    <a href="https://gnomad.broadinstitute.org/gene/${scores['g_name']}?dataset=${gnomadVersion}" class="small-link" target="_blank">gnomAD</a>,
                    <a href="https://search.clinicalgenome.org/kb/genes?page=1&size=25&search=${scores['g_name']}" class="small-link" target="_blank">ClinGen</a>,
                    <a href="https://useast.ensembl.org/Homo_sapiens/Gene/Summary?g=${scores['g_name']}" class="small-link" target="_blank">Ensembl</a>,
                    <a href="https://www.deciphergenomics.org/gene/${scores['g_name']}" class="small-link" target="_blank">Decipher</a>,
                    <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=${scores['g_name']}" class="small-link" target="_blank">GeneCards</a>
                </td>`

            for (const [i, label, scoreKey, positionKey, REF_scoreKey, ALT_scoreKey] of (
                isPangolin ? [
                    [0, "Splice Loss", "DS_SL", "DP_SL", "SL_REF", "SL_ALT"],
                    [1, "Splice Gain", "DS_SG", "DP_SG", "SG_REF", "SG_ALT"],
                ] : [
                    [0, "Acceptor Loss", "DS_AL", "DP_AL", "DS_AL_REF", "DS_AL_ALT"],
                    [1, "Donor Loss",    "DS_DL", "DP_DL", "DS_DL_REF", "DS_DL_ALT"],
                    [2, "Acceptor Gain", "DS_AG", "DP_AG", "DS_AG_REF", "DS_AG_ALT"],
                    [3, "Donor Gain",    "DS_DG", "DP_DG", "DS_DG_REF", "DS_DG_ALT"],
                ])) {
                if (i > 0) {
                    row += `</tr><tr class="${resultRowClasses.join(' ')}">`
                }

                const modalDialogId = transcriptIndex*10 + i
                const detailedScoresTable = generateTableOfScoresForInsertedBases(
                    modalDialogId, scores[scoreKey], scores[positionKey], ref, alt, scores["SCORES_FOR_INSERTED_BASES"])

                if (detailedScoresTable) {
                    modalDialogIds.push(modalDialogId)
                }

                row += `<td>${label}</td>
                        <td ${getScoreStyle(scores[scoreKey])}>
                            ${formatScore(scores[scoreKey])}
                            ${detailedScoresTable}
                         </td>
                        <td>${
                            (
                                parseFloat(scores[scoreKey]) == 0 &&
                                parseFloat(scores[REF_scoreKey]) == 0 &&
                                parseFloat(scores[ALT_scoreKey]) == 0
                            )? "" : updatePositionAccountingForInsertedBases(scoreKey, scores[scoreKey], scores[positionKey], ref, alt, scores["SCORES_FOR_INSERTED_BASES"])}
                        </td>`

                if (showRefAltScoreColumns == "1") {
                    row += `<td class="ref-score-column">${formatScore(scores[REF_scoreKey])}</td>
                            <td class="alt-score-column">${formatScore(scores[ALT_scoreKey])}</td>`
                }
            }
            row += "</tr>"

            tableRows.push(row)
            transcriptIndex++
        }

        $(`#${tool.toLowerCase()}-header`).after(tableRows.join(""))

        if (transcriptCategories["MS"]) {
            $(".main-transcript-label").html("MANE Select")
        } else if (transcriptCategories["MP"]) {
            $(".main-transcript-label").html("MANE Plus Clinical")
        } else if (transcriptCategories["C"]) {
            $(".main-transcript-label").html("Canonical")
        } else {
            $("#transcript-button-table").hide()
        }

        // initialize any modal dialogs
        for (const modalDialogIndex of modalDialogIds) {
            $(`#table-icon${modalDialogIndex}`).click(() => {
                $(`#modal-with-table${modalDialogIndex}`).modal('show')
            })
        }
        $("#transcript-button-table").show()
        updateTranscriptButtons(Object.keys(transcriptCategories).length > 1 ? "main" : "all")

        return apiResponseJson
    }

    const updateTranscriptButtons = (category) => {
        $("#main-transcript-button, #all-transcript-button").removeClass("primary")
        $(`#${category}-transcript-button`).addClass("primary")

        if (category == "main") {
            $(".spliceai-result-row").hide()
            $(".spliceai-result-row.main-transcript").show()
        } else if (category == "all") {
            $(".spliceai-result-row").show()
        }
    }

    const updateVisualizationCheckboxes = (readFromLocalStorage, genomeVersion) => {
        /** 
         * Retrieve and save the current state of checkboxes in the Visualization section, or set the state of these 
         * checkboxes based on cookies / local storage.
         */
        let tracksToShow
        if (readFromLocalStorage) {
            // default tracks
            tracksToShow = {
                "igv-variant": true,
                "igv-spliceai-ref-alt": true,
                "igv-spliceai-delta-scores": true,
                "igv-pangolin-delta-scores": true,
            }

            // if there are tracksToShow in local storage, overwrite defaults with those settings
            const fromStorage = localStorage.getItem("tracksToShow")
            if (fromStorage) {
                const fromStorageJson = JSON.parse(fromStorage)
                for (const key of Object.keys(fromStorageJson)) {
                    tracksToShow[key] = fromStorageJson[key]
                    try {
                        $(`input[name='${key}']`).prop( "checked", tracksToShow[key] )
                    } catch (e) {
                        console.log(e)
                    }
                }
            }
        } else {

            // retrieve state from checkboxes
            tracksToShow = {}
            for (const name of ["igv-variant", "igv-spliceai-ref-alt", "igv-spliceai-delta-scores", "igv-pangolin-delta-scores", "igv-gencode-genes"]) {
                tracksToShow[name] = $(`input[name='${name}']`).prop("checked")
            }
            for (const name of ["igv-100-mer-mappability", "igv-segdups"]) {
                // these tracks are not available for hg38
                $(`input[name='${name}']`).prop("checked", $(`input[name='${name}']`).prop("checked") && genomeVersion != "37")
                $(`input[name='${name}']`).prop("disabled", genomeVersion == "37")

                tracksToShow[name] = $(`input[name='${name}']`).prop("checked")
            }
            for (const minScore of [0.5, 0.2]) {   //0.2,  hide the >= 0.2 tracks for now
                for (const splicePredictionType of ["loss", "gain"]) {
                    const name = `igv-spliceai-precomputed-${splicePredictionType}-${minScore}`

                    // these tracks are not available for hg19
                    $(`input[name='${name}']`).prop("checked", $(`input[name='${name}']`).prop("checked") && genomeVersion != "37")
                    $(`input[name='${name}']`).prop("disabled", genomeVersion == "37")

                    tracksToShow[name] = $(`input[name='${name}']`).prop("checked")
                }
            }
            for (const tissue of ["blood", "fibroblasts", "muscle", "lymphocytes", "brain-cortex"]) {
                // these tracks are not available for hg19
                $(`input[name='igv-gtex-${tissue}']`).prop("checked", $(`input[name='igv-gtex-${tissue}']`).prop("checked") && genomeVersion != "37")
                $(`input[name='igv-gtex-${tissue}']`).prop("disabled", genomeVersion == "37")
                $(`input[name='igv-gtex-${tissue}-all']`).prop("checked", $(`input[name='igv-gtex-${tissue}-all']`).prop("checked") && genomeVersion != "37")
                $(`input[name='igv-gtex-${tissue}-all']`).prop("disabled", genomeVersion == "37")

                tracksToShow[`igv-gtex-${tissue}`] = $(`input[name='igv-gtex-${tissue}']`).prop("checked")
                tracksToShow[`igv-gtex-${tissue}-all`] = $(`input[name='igv-gtex-${tissue}-all']`).prop("checked")
            }
            //console.log("Set local storage tracksToShow", tracksToShow)
            localStorage.setItem("tracksToShow", JSON.stringify(tracksToShow))
        }
        
        return tracksToShow
    }
    
    
    const generateIgvConfig = (spliceaiResponseJson, pangolinResponseJson, genomeVersion) => {
        /* Generates an igv.js config json objects based on the predicted scores from SpliceAI and/or Pangolin
        *
        * Args:
        *   allNonZeroScoresFromSpliceAI (array): an array of acceptor/donor loss/gain scores from SpliceAI
        *   allNonZeroScoresFromPangolin (array): an array of acceptor/donor loss/gain scores from Pangolin
        *   genomeVersion (string): "37" or "38"
        *
        * Return:
        *   object: an igv.js config json object
        */

        const apiResponseJson = spliceaiResponseJson || pangolinResponseJson
        let chrom = apiResponseJson.chrom
        chrom = `chr${chrom.replace('chr', '')}`
        const variantPos = apiResponseJson.pos
        const variantRef = apiResponseJson.ref
        const variantAlt = apiResponseJson.alt

        const tracksToShow = updateVisualizationCheckboxes(false, genomeVersion)

        // log event
        makeRequest(`${baseApiUrl['pangolin-37']}/log/show_igv?details=${encodeURIComponent(JSON.stringify(tracksToShow))}&hg=${genomeVersion}&distance=${apiResponseJson.distance}&mask=${apiResponseJson.mask}&variant=${apiResponseJson.variant}`)

        let minPos = 1e9
        let maxPos = 0
        for (const apiResponse of [spliceaiResponseJson, pangolinResponseJson]) {
            if (!apiResponse) {
                continue
            }
            if (!apiResponse.allNonZeroScores) {
                continue
            }
            for (const scores of apiResponse.allNonZeroScores) {
                scores["chr"] = chrom
                scores["start"] = scores["pos"]
                scores["end"] = scores["pos"]
                minPos = Math.min(minPos, scores["pos"])
                maxPos = Math.max(maxPos, scores["pos"])
            }
        }

        // specify IGV tracks and the data to display in them
        const tracks = []

        tracks.push({
            name: "Refseq",
            format: "refgene",
            url: genomeVersion == "38" ?
                "https://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/ncbiRefSeq.txt.gz" :
                "https://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/ncbiRefSeq.txt.gz",
            indexed: false,
            infoURL: "https://www.ncbi.nlm.nih.gov/gene/?term=$$",
            height: 100,
        })

        if (tracksToShow["igv-variant"]) {
            tracks.push({
                name: "Variant",
                type: "vcf",
                description: `This track shows the location of the <b>${chrom}:${variantPos} ${variantRef}>${variantAlt}</b> variant`,
                height: 30,
                features: [
                    {
                        chr: chrom,   //see createVCFVariant function in the igv.js repo for the allowed fields
                        pos: variantPos,
                        start: variantPos - 1,
                        end: variantPos + variantRef.length - 1,
                        referenceAllele: variantRef,
                        alternateBases: variantAlt,
                        names: ".", // id in VCF
                        info: {
                            variant: `${chrom}-${variantPos}-${variantRef}-${variantAlt}`,
                        }
                    },
                ],
            })
        }

        if (spliceaiResponseJson) {
            if (tracksToShow["igv-spliceai-ref-alt"]) {
                tracks.push({
                    name: `SpliceAI REF/ALT`,
                    description: `This track shows SpliceAI scores for the
                                    <span style="color:#0000B4"><b>reference sequence</b></span> (without the variant) and the
                                    <span style="color:#05d0d2"><b>alternate sequence</b></span> (with the variant). <br />
                                    An <b>A</b> (acceptor) or <b>D</b> (donor) symbol is shown where SpliceAI predicts there to be a splice donor or acceptor with score ≥ 0.01. <br/>
                                    These symbols are shown up-side-down when the score for the alternate sequence is lower than the score for the reference sequence.
                                    Numberical labels represent scores for the reference sequence. These predictions are based on the pre-mRNA sequence for transcript <br />
                                    <b>${spliceaiResponseJson.allNonZeroScoresTranscriptId}</b> (${spliceaiResponseJson.allNonZeroScoresStrand == '-' ? 'minus' : 'plus'} strand)
                                    within a +/- ${spliceaiResponseJson.distance}bp window around <b>${chrom}:${variantPos} ${variantRef}>${variantAlt}</b>.`,
                    height: 100,
                    rawOrDelta: "raw",
                    tool: "SpliceAI",
                    type: "spliceprediction",
                    features: spliceaiResponseJson.allNonZeroScores || [],
                    strand: spliceaiResponseJson.allNonZeroScoresStrand || "+",
                })
            }
            if (tracksToShow["igv-spliceai-delta-scores"]) {
                tracks.push({
                    name: `SpliceAI Δ`,
                    description: `This track shows <b>A</b> (acceptor) and <b>D</b> (donor) symbols at positions where SpliceAI predicts a delta score ≥ 0.01. <br/>
                                  Vertical lines are <b style="color:#FF0000">red</b> for delta scores ≥ 0.8,
                                  <b style="color:#FFCB1F">yellow</b> for delta scores ≥ 0.5, and
                                  <b style="color:#1fb839">green</b> for delta scores ≥ 0.2<br />
                                  These predictions are based on the pre-mRNA sequence for transcript
                                  <b>${spliceaiResponseJson.allNonZeroScoresTranscriptId}</b> (${spliceaiResponseJson.allNonZeroScoresStrand == '-' ? 'minus' : 'plus'} strand)<br />
                                  within a +/- ${spliceaiResponseJson.distance}bp window around <b>${chrom}:${variantPos} ${variantRef}>${variantAlt}</b>.`,
                    height: 200,
                    rawOrDelta: "delta",
                    tool: "SpliceAI",
                    type: "spliceprediction",
                    features: spliceaiResponseJson.allNonZeroScores || [],
                    strand: spliceaiResponseJson.allNonZeroScoresStrand || "+",
                })
            }
        }

        if (pangolinResponseJson) {
            if (tracksToShow["igv-pangolin-delta-scores"]) {

                tracks.push({
                    name: `Pangolin Δ`,
                    description: `This track shows a <b>P</b> at positions where Pangolin predicts a delta score ≥ 0.01. <br/>
                                  Vertical lines are <b style="color:#FF0000">red</b> for delta scores ≥ 0.8,
                                  <b style="color:#FFCB1F">yellow</b> for delta scores ≥ 0.5, and
                                  <b style="color:#1fb839">green</b> for delta scores ≥ 0.2<br />
                                  These predictions are based on the pre-mRNA sequence for transcript
                                  <b>${pangolinResponseJson.allNonZeroScoresTranscriptId}</b> (${pangolinResponseJson.allNonZeroScoresStrand == '-' ? 'minus' : 'plus'} strand)<br />
                                  within a +/- ${pangolinResponseJson.distance}bp window around <b>${chrom}:${variantPos} ${variantRef}>${variantAlt}</b>.`,
                    height: 200,
                    rawOrDelta: "delta",
                    tool: "Pangolin",
                    type: "spliceprediction",
                    features: pangolinResponseJson.allNonZeroScores || [],
                    strand: pangolinResponseJson.allNonZeroScoresStrand || "+",
                })
            }
        }

        if (tracksToShow["igv-gencode-genes"]) {
            const gencodeTrackPath = `gs://tgg-viewer/ref/GRCh${genomeVersion}/gencode_${GENCODE_VERSION}/gencode.${GENCODE_VERSION}.GRCh${genomeVersion}.sorted.txt.gz`

            tracks.push({
                name: `Gencode ${GENCODE_VERSION}`,
                format: 'refgene',
                url: gencodeTrackPath,
                indexURL: `${gencodeTrackPath}.tbi`,
                indexed: true,
                searchable: true,
                height: 350,
                visibilityWindow: -1,
                order: 1000001,
                displayMode: 'EXPANDED',
                color: 'rgb(76,171,225)',
            })
        }

        if (genomeVersion == "38") {
            for (const minScore of [0.5, 0.2]) {   //0.2,  hide the >= 0.2 tracks for now
                for (const splicePredictionType of ["loss", "gain"]) {
                    if (tracksToShow[`igv-spliceai-precomputed-${splicePredictionType}-${minScore}`]) {
                        tracks.push(
                            {
                                name: `SpliceAI: A or D ${splicePredictionType} ≥ ${minScore}`,
                                description: `This track visualizes Illumina's precomputed SpliceAI score tables for SNVs and small INDELs.<br/>
                                     Each genomic location of an SNV or INDEL variant with a SpliceAI &#916; score ≥ ${minScore} is shown as the origin of an arrow.<br />
                                     The arrow points to the location where that variant would likely cause acceptor or donor ${splicePredictionType}.
                                     Clicking on the arrow displays additional details.`,
                                type: "spliceJunctions",
                                height: 100,
                                url: `gs://tgg-viewer/ref/GRCh38/spliceai/spliceai_scores.raw.snps_and_indels.hg38.filtered.sorted.score_${minScore}.splice_${splicePredictionType}.bed.gz`,
                                indexURL: `gs://tgg-viewer/ref/GRCh38/spliceai/spliceai_scores.raw.snps_and_indels.hg38.filtered.sorted.score_${minScore}.splice_${splicePredictionType}.bed.gz.tbi`,
                            }
                        )
                    }
                }
            }


            for (const [filenamePrefix, tissue, sampleCount] of [
                ["GTEX_blood.755_samples", "blood", 755],
                ["GTEX_fibs.504_samples", "fibroblasts", 504],
                ["GTEX_muscle.803_samples", "muscle", 803],
                ["GTEX_lymphocytes.174_samples", "lymphocytes", 174],
                ["GTEX_brain_cortex.255_samples", "brain-cortex", 255],
                //["GTEX_frontal_cortex.209_samples", "frontal cortex", 803],
            ]) {
                for (const normalized of [true, false]) {
                    if (!tracksToShow[`igv-gtex-${tissue}${normalized ? '' : '-all'}`]) {
                        continue
                    }

                    const filenameSuffix = normalized ? ".normalized" : ""
                    tracks.push({
                        name: `GTEx ${tissue}: ${normalized ? 'per-sample average' : `summed over all ${sampleCount} samples`}`,
                        description: `This track shows the combined splice junctions from all ${sampleCount} ${tissue} samples available in GTEx v8.
                                      Splice junctions are labeled with the total number of RNA-seq reads that supported the junction, summed across the ${sampleCount} samples${normalized ? ' and divided by ' + sampleCount : ''}.
                                      The coverage track shows the total number of RNA-seq reads that overlap each position, summed across all samples${normalized ? ' and divided by ' + sampleCount : ''}.`,
                        type: "merged",
                        height: 100,
                        tracks: [{
                            type: "wig",
                            format: "bigwig",
                            url: `gs://tgg-viewer/ref/GRCh38/gtex_v8/${filenamePrefix}${filenameSuffix}.bigWig`,
                        }, {
                            type: 'spliceJunctions',
                            format: 'bed',
                            minJunctionEndsVisible: 1,
                            colorBy: 'strand',
                            minTotalReads: 3,
                            url: `gs://tgg-viewer/ref/GRCh38/gtex_v8/${filenamePrefix}${filenameSuffix}.junctions.bed.gz`,
                            indexURL: `gs://tgg-viewer/ref/GRCh38/gtex_v8/${filenamePrefix}${filenameSuffix}.junctions.bed.gz.tbi`,
                        }],
                    })
                }
            }

            if (tracksToShow["igv-100-mer-mappability"]) {
                tracks.push({
                    name: "100-mer mappability",
                    description: "100bp k-mer mappability track from UCSC. Lower values indicate regions that are not unique",
                    type: "wig",
                    format: "bigwig",
                    url: "gs://tgg-viewer/ref/GRCh38/mappability/GRCh38_no_alt_analysis_set_GCA_000001405.15-k100_m2.bw",
                    height: 100,
                })
            }

            if (tracksToShow["igv-segdups"]) {
                tracks.push({
                    name: "SegDups",
                    description: "Segmental duplications track from UCSC",
                    format: "gtf",
                    url: "gs://tgg-viewer/ref/GRCh38/segdups/segdups.gtf.gz",
                    indexURL: "gs://tgg-viewer/ref/GRCh38/segdups/segdups.gtf.gz.tbi",
                    height: 100,
                })
            }
        }

        let locusMargin = maxPos - minPos < 200 ? 15 : 50

        // igv.js reference genome specs are copied from https://igv.org/genomes/genomes.json
        // These must be specified explicitly rather than just setting the reference to "hg19" or "hg38" so that
        // the RefSeq gene track can be the first track (as described in https://github.com/igvteam/igv.js/issues/1518)
        let reference = genomeVersion == "37" ? {
            "id": "hg19",
            "name": "Human (GRCh37/hg19)",
            "fastaURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg19/hg19.fasta",
            "indexURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg19/hg19.fasta.fai",
            "cytobandURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg19/cytoBand.txt",
            "aliasURL": "https://s3.amazonaws.com/igv.org.genomes/hg19/hg19_alias.tab",
            "chromosomeOrder": "chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10, chr11, chr12, chr13, chr14, chr15, chr16, chr17, chr18, chr19, chr20, chr21, chr22, chrX, chrY",
            "tracks": tracks,
        } : {
            "id": "hg38",
            "name": "Human (GRCh38/hg38)",
            "fastaURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg38/hg38.fa",
            "indexURL": "https://igv-genepattern-org.s3.amazonaws.com/genomes/seq/hg38/hg38.fa.fai",
            "cytobandURL": "https://s3.amazonaws.com/igv.org.genomes/hg38/annotations/cytoBandIdeo.txt.gz",
            "aliasURL": "https://s3.amazonaws.com/igv.org.genomes/hg38/hg38_alias.tab",
            "chromosomeOrder": "chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10, chr11, chr12, chr13, chr14, chr15, chr16, chr17, chr18, chr19, chr20, chr21, chr22, chrX, chrY",
            "tracks": tracks,
        }

        return {
            reference: reference,
            showCursorTrackingGuide: true,
            locus: `${chrom}:${minPos - locusMargin}-${maxPos + locusMargin}`,
        }
    }

    const updateIgvBrowser = async (spliceaiResponseJson, pangolinResponseJson, genomeVersion) => {
        if ((spliceaiResponseJson && spliceaiResponseJson.allNonZeroScores) || (
            pangolinResponseJson && pangolinResponseJson.allNonZeroScores)) {
            const igvConfig = generateIgvConfig(spliceaiResponseJson, pangolinResponseJson, genomeVersion)
            if (!window.igvBrowser) {
                // create igv.js browser object if it hasn't been created yet
                window.igvBrowser = await igv.createBrowser(document.getElementById("igv-viewport"), igvConfig)
            } else {
                console.log("Updating igv config to", igvConfig)
                await window.igvBrowser.loadSessionObject(igvConfig)
            }
        }
    }


    const handleSubmit = async () => {
        // get user inputs from form elements
        const variant = $("#search-box").val().trim()
        if (!variant) return
        const genomeVersion = $("input[name='hg']:checked").val().trim()
        const basicOrComprehensive = $("input[name='gencode-gene-set']:checked").val().trim()
        const maxDistance = $("#max-distance-input").val().trim()
        const mask = $(`input[name='mask']`).prop("checked") ? "1" : "0";
        const showRefAltColumns = $(`input[name='show-ref-alt']`).prop("checked") ? "1" : "0";

        updateVisualizationCheckboxes(false, genomeVersion)

        //start processing submit
        $("#submit-button").addClass(["loading", "disabled"])
        $("#response-box, #spliceai-table, #pangolin-table, #transcript-button-table, #error-box").hide()
        $("#error-box").html("")

        $(".ref-score-column, .alt-score-column").toggle(showRefAltColumns == "1")


        // make API calls
        try {
            const normalizedVariantAndConsequence = await normalizeVariant(variant, genomeVersion)
            const normalizedVariant = normalizedVariantAndConsequence.variant
            const variantConsequence = normalizedVariantAndConsequence.consequence

            //let [spliceaiResponseJson, pangolinResponseJson, variantAnnotationsJson] = await Promise.all([
            let [spliceaiResponseJson, pangolinResponseJson] = await Promise.all([
                generateSplicingResultsTable(normalizedVariant, variantConsequence, "SpliceAI", variant, genomeVersion, basicOrComprehensive, maxDistance, mask, showRefAltColumns),
                generateSplicingResultsTable(normalizedVariant, variantConsequence, "Pangolin", variant, genomeVersion, basicOrComprehensive, maxDistance, mask, showRefAltColumns),
                //generateVariantAnnotationTable(normalizedVariant, genomeVersion),
            ].map(p => p.catch(e => ({'error': e.message}))))  // if an error is thrown, return {'error': e} instead of the JSON response

            $("#response-box").show()

            $(".question, .exclamation, .score-table").popup({
                "on": "click",
                "lastResort": "bottom left",
            })
            $(".question").css({"color": "#666666", "cursor": "pointer"})
            $(".exclamation").css({"cursor": "pointer"})
            $(".score-table").css({"color": "#000080", "cursor": "pointer"})
            $(".ui.modal").modal()

            if (spliceaiResponseJson.error) {
                showError(`${spliceaiResponseJson.error}`)
                spliceaiResponseJson = null
                $("#spliceai-table, #transcript-button-table").hide()
            } else {
                $("#spliceai-table, #transcript-button-table").show()
            }

            if (pangolinResponseJson.error) {
                showError(`${pangolinResponseJson.error}`)
                pangolinResponseJson = null
                $("#pangolin-table").hide()
            } else {
                $("#pangolin-table").show()
            }

            // update url params
            window.location.hash = "#" + $.param({
                variant: variant,
                hg: genomeVersion,
                bc: basicOrComprehensive,
                distance: maxDistance,
                mask: mask,
                ra: showRefAltColumns,
            })

            lastGenomeVersion = genomeVersion
            lastSpliceaiResponseJson = spliceaiResponseJson
            lastPangolinResponseJson = pangolinResponseJson

            if (spliceaiResponseJson != null || pangolinResponseJson != null) {
                $("#igv-table").show()
                $("#igv-div").hide()
                $("#show-igv-button").text("Show")
            }
        } catch(e) {
            console.error(e)
            showError(e.message)
        }

        //done processing submit
        $("#submit-button").removeClass(["loading", "disabled"])
    }

    const applyUrlSettingsToFormElements = () => {
        // get optional settings from the url and update form settings
        let hgFromUrl = $.urlParam('hg')
        if (hgFromUrl == "19") {
            hgFromUrl = "37"
        }
        if (hgFromUrl && (hgFromUrl == "38" || hgFromUrl == "37")) {
            $(`input[name='hg'][value='${hgFromUrl}']`).prop("checked", true)
        }

        const bcFromUrl = $.urlParam('bc')
        if (bcFromUrl && (bcFromUrl == "basic" || bcFromUrl == "comprehensive")) {
            $(`input[name='gencode-gene-set'][value='${bcFromUrl}']`).prop("checked", true)
        }

        const maxDistanceFromUrl = $.urlParam('distance')
        if (maxDistanceFromUrl) {
            $("#max-distance-input").val(maxDistanceFromUrl)
        }

        const maskFromUrl = $.urlParam('mask')
        if (maskFromUrl) {
            $(`input[name='mask']`).prop("checked", maskFromUrl == "1")
        }

        const showRefAltColumnsFromUrl = $.urlParam('ra')
        if (showRefAltColumnsFromUrl) {
            $(`input[name='show-ref-alt']`).prop("checked", showRefAltColumnsFromUrl == "1")
        }

        //update the variant input box last and trigger a search
        const variantFromUrl = $.urlParam('variant')
        if (variantFromUrl) {
            $("#search-box").val(variantFromUrl)
            $("#submit-button").click()
        }
    }

    /*
    const toggleRefAltScoreColumns = () => {
        const currentText = $("#toggle-ref-alt-scores-button").html()

        if (currentText.match(/Show/i)) {
            $(".ref-alt-score-column").show()
            $("#toggle-ref-alt-scores-button").removeClass('primary')
            $("#toggle-ref-alt-scores-button").html('Hide REF & ALT Scores')
        } else {
            $(".ref-alt-score-column").hide()
            $("#toggle-ref-alt-scores-button").addClass('primary')
            $("#toggle-ref-alt-scores-button").html('Show REF & ALT Scores')
        }
    }
    */

    // define function for parsing url parameters like ?variant=chr1-1234567-T-G
    $.urlParam = (name) => {
        const results = new RegExp('[\?&#]?' + name + '=([^&#]*)').exec(window.location.hash);
        return (results !== null) ? decodeURIComponent(results[1]) || 0 : false;
    }

    $(document).ready(() => {
        $("#gencode-version").text(GENCODE_VERSION)
        // init UI elements
        $(".ui.checkbox").checkbox()
        $(".question, .exclamation, .score-table").popup({"on": "click"})
        $(".question, .exclamation, .score-table").css("cursor", "pointer")
        $("#search-box").focus()
        $("#response-box").hide()

        // init event handlers
        $("#submit-button").click(handleSubmit)

        $("#search-box, #max-distance-input").keydown((event) => {
            if ((event.keyCode || event.which) == 13) {
                $("#submit-button").click()
            }
        })

        $("#show-igv-button").click(async () => {
            $("#show-igv-button").text("Update")
            $("#igv-div").show()
            await updateIgvBrowser(lastSpliceaiResponseJson, lastPangolinResponseJson, lastGenomeVersion)
        })

        updateVisualizationCheckboxes(true, "38")

        applyUrlSettingsToFormElements()
    })

</script>
</body>
</html>
